From 2adc33edd06a945cacfc7c25916ef41202ac6fbb Mon Sep 17 00:00:00 2001
From: Li Wei <liwei825@huawei.com>
Date: Wed, 15 Feb 2023 16:44:35 +0800
Subject: [PATCH 322/921] [Huawei] ocfs2: add customized modifications for
 lockglue

Offering: HULK
ocfs2 inclusion
category: feature
bugzilla: NA

----------------------------------------

Add customized modifications for the additional of lockglue.

Signed-off-by: Li Wei <liwei825@huawei.com>
---
 fs/ocfs2/acl.c          |   6 +-
 fs/ocfs2/dcache.c       |  58 +++--
 fs/ocfs2/dir.c          | 130 ++++++----
 fs/ocfs2/export.c       |  10 +-
 fs/ocfs2/extent_map.c   |  27 ++-
 fs/ocfs2/extent_map.h   |   3 +
 fs/ocfs2/ioctl.c        | 178 +++++++++++++-
 fs/ocfs2/mmap.c         |   9 +-
 fs/ocfs2/move_extents.c |  16 +-
 fs/ocfs2/namei.c        | 513 ++++++++++++++++++++++++++++++++++------
 fs/ocfs2/namei.h        |   2 +
 fs/ocfs2/quota_global.c |  18 +-
 fs/ocfs2/quota_local.c  |  40 ++--
 fs/ocfs2/refcounttree.c |  52 ++--
 fs/ocfs2/xattr.c        | 273 ++++++++++++++++++---
 fs/ocfs2/xattr.h        |  18 ++
 16 files changed, 1092 insertions(+), 261 deletions(-)

diff --git a/fs/ocfs2/acl.c b/fs/ocfs2/acl.c
index 7b07f5df3a29..db7da0d25f86 100644
--- a/fs/ocfs2/acl.c
+++ b/fs/ocfs2/acl.c
@@ -20,7 +20,7 @@
 
 #include "ocfs2.h"
 #include "alloc.h"
-#include "dlmglue.h"
+#include "lockglue.h"
 #include "file.h"
 #include "inode.h"
 #include "journal.h"
@@ -266,7 +266,7 @@ int ocfs2_iop_set_acl(struct inode *inode, struct posix_acl *acl, int type)
 {
 	struct buffer_head *bh = NULL;
 	int status, had_lock;
-	struct ocfs2_lock_holder oh;
+	struct ocfs2_lock_holder *oh;
 
 	had_lock = ocfs2_inode_lock_tracker(inode, &bh, 1, &oh);
 	if (had_lock < 0)
@@ -295,7 +295,7 @@ struct posix_acl *ocfs2_iop_get_acl(struct inode *inode, int type)
 	struct buffer_head *di_bh = NULL;
 	struct posix_acl *acl;
 	int had_lock;
-	struct ocfs2_lock_holder oh;
+	struct ocfs2_lock_holder *oh;
 
 	osb = OCFS2_SB(inode->i_sb);
 	if (!(osb->s_mount_opt & OCFS2_MOUNT_POSIX_ACL))
diff --git a/fs/ocfs2/dcache.c b/fs/ocfs2/dcache.c
index 42a61eecdacd..b9c44190235b 100644
--- a/fs/ocfs2/dcache.c
+++ b/fs/ocfs2/dcache.c
@@ -20,7 +20,7 @@
 
 #include "alloc.h"
 #include "dcache.h"
-#include "dlmglue.h"
+#include "lockglue.h"
 #include "file.h"
 #include "inode.h"
 #include "ocfs2_trace.h"
@@ -33,7 +33,6 @@ void ocfs2_dentry_attach_gen(struct dentry *dentry)
 	dentry->d_fsdata = (void *)gen;
 }
 
-
 static int ocfs2_dentry_revalidate(struct dentry *dentry, unsigned int flags)
 {
 	struct inode *inode;
@@ -93,6 +92,9 @@ static int ocfs2_dentry_revalidate(struct dentry *dentry, unsigned int flags)
 		goto bail;
 	}
 
+	if (ocfs2_disk_lock(osb))
+		goto valid;
+
 	/*
 	 * If the last lookup failed to create dentry lock, let us
 	 * redo it.
@@ -216,6 +218,8 @@ int ocfs2_dentry_attach_lock(struct dentry *dentry,
 	int ret;
 	struct dentry *alias;
 	struct ocfs2_dentry_lock *dl = dentry->d_fsdata;
+	struct ocfs2_super *osb;
+	unsigned long flags;
 
 	trace_ocfs2_dentry_attach_lock(dentry->d_name.len, dentry->d_name.name,
 				       (unsigned long long)parent_blkno, dl);
@@ -229,6 +233,10 @@ int ocfs2_dentry_attach_lock(struct dentry *dentry,
 	if (!inode)
 		return 0;
 
+	osb = OCFS2_SB(dentry->d_sb);
+	if (ocfs2_disk_lock(osb))
+		return 0;
+
 	if (d_really_is_negative(dentry) && dentry->d_fsdata) {
 		/* Converting a negative dentry to positive
 		   Clear dentry->d_fsdata */
@@ -295,7 +303,7 @@ int ocfs2_dentry_attach_lock(struct dentry *dentry,
 	ocfs2_dentry_lock_res_init(dl, parent_blkno, inode);
 
 out_attach:
-	spin_lock(&dentry_attach_lock);
+	spin_lock_irqsave(&dentry_attach_lock, flags);
 	if (unlikely(dentry->d_fsdata && !alias)) {
 		/* d_fsdata is set by a racing thread which is doing
 		 * the same thing as this thread is doing. Leave the racing
@@ -303,14 +311,14 @@ int ocfs2_dentry_attach_lock(struct dentry *dentry,
 		 */
 		spin_unlock(&dentry_attach_lock);
 		iput(dl->dl_inode);
-		ocfs2_lock_res_free(&dl->dl_lockres);
+		ocfs2_lock_res_free(osb, &dl->dl_lockres);
 		kfree(dl);
 		return 0;
 	}
 
 	dentry->d_fsdata = dl;
 	dl->dl_count++;
-	spin_unlock(&dentry_attach_lock);
+	spin_unlock_irqrestore(&dentry_attach_lock, flags);
 
 	/*
 	 * This actually gets us our PRMODE level lock. From now on,
@@ -329,11 +337,11 @@ int ocfs2_dentry_attach_lock(struct dentry *dentry,
 	 * which means iput() will not be called during dput(dentry).
 	 */
 	if (ret < 0 && !alias) {
-		ocfs2_lock_res_free(&dl->dl_lockres);
+		ocfs2_lock_res_free(osb, &dl->dl_lockres);
 		BUG_ON(dl->dl_count != 1);
-		spin_lock(&dentry_attach_lock);
+		spin_lock_irqsave(&dentry_attach_lock, flags);
 		dentry->d_fsdata = NULL;
-		spin_unlock(&dentry_attach_lock);
+		spin_unlock_irqrestore(&dentry_attach_lock, flags);
 		kfree(dl);
 		iput(inode);
 	}
@@ -367,9 +375,10 @@ int ocfs2_dentry_attach_lock(struct dentry *dentry,
 static void ocfs2_drop_dentry_lock(struct ocfs2_super *osb,
 				   struct ocfs2_dentry_lock *dl)
 {
-	iput(dl->dl_inode);
 	ocfs2_simple_drop_lockres(osb, &dl->dl_lockres);
-	ocfs2_lock_res_free(&dl->dl_lockres);
+	ocfs2_lock_res_free(osb, &dl->dl_lockres);
+
+	iput(dl->dl_inode);
 	kfree(dl);
 }
 
@@ -377,13 +386,14 @@ void ocfs2_dentry_lock_put(struct ocfs2_super *osb,
 			   struct ocfs2_dentry_lock *dl)
 {
 	int unlock = 0;
+	unsigned long flags;
 
 	BUG_ON(dl->dl_count == 0);
 
-	spin_lock(&dentry_attach_lock);
+	spin_lock_irqsave(&dentry_attach_lock, flags);
 	dl->dl_count--;
 	unlock = !dl->dl_count;
-	spin_unlock(&dentry_attach_lock);
+	spin_unlock_irqrestore(&dentry_attach_lock, flags);
 
 	if (unlock)
 		ocfs2_drop_dentry_lock(osb, dl);
@@ -391,8 +401,13 @@ void ocfs2_dentry_lock_put(struct ocfs2_super *osb,
 
 static void ocfs2_dentry_iput(struct dentry *dentry, struct inode *inode)
 {
-	struct ocfs2_dentry_lock *dl = dentry->d_fsdata;
+	struct ocfs2_dentry_lock *dl;
+	struct ocfs2_super *osb = OCFS2_SB(dentry->d_sb);
 
+	if (ocfs2_disk_lock(osb))
+		goto out;
+
+	dl = dentry->d_fsdata;
 	if (!dl) {
 		/*
 		 * No dentry lock is ok if we're disconnected or
@@ -403,9 +418,9 @@ static void ocfs2_dentry_iput(struct dentry *dentry, struct inode *inode)
 			unsigned long long ino = 0ULL;
 			if (inode)
 				ino = (unsigned long long)OCFS2_I(inode)->ip_blkno;
-			mlog(ML_ERROR, "Dentry is missing cluster lock. "
+			mlog(ML_ERROR, "%s: Dentry is missing cluster lock. "
 			     "inode: %llu, d_flags: 0x%x, d_name: %pd\n",
-			     ino, dentry->d_flags, dentry);
+			     osb->uuid_str, ino, dentry->d_flags, dentry);
 		}
 
 		goto out;
@@ -414,7 +429,7 @@ static void ocfs2_dentry_iput(struct dentry *dentry, struct inode *inode)
 	mlog_bug_on_msg(dl->dl_count == 0, "dentry: %pd, count: %u\n",
 			dentry, dl->dl_count);
 
-	ocfs2_dentry_lock_put(OCFS2_SB(dentry->d_sb), dl);
+	ocfs2_dentry_lock_put(osb, dl);
 
 out:
 	iput(inode);
@@ -455,6 +470,9 @@ void ocfs2_dentry_move(struct dentry *dentry, struct dentry *target,
 	if (old_dir == new_dir)
 		goto out_move;
 
+	if (ocfs2_disk_lock(osb))
+		goto out_move;
+
 	ocfs2_dentry_lock_put(osb, dentry->d_fsdata);
 
 	dentry->d_fsdata = NULL;
@@ -466,7 +484,15 @@ void ocfs2_dentry_move(struct dentry *dentry, struct dentry *target,
 	d_move(dentry, target);
 }
 
+int ocfs2_dentry_delete(const struct dentry *dentry)
+{
+	if (ocfs2_disk_lock(OCFS2_SB(dentry->d_sb)))
+		return 1;
+	return 0;
+}
+
 const struct dentry_operations ocfs2_dentry_ops = {
 	.d_revalidate		= ocfs2_dentry_revalidate,
 	.d_iput			= ocfs2_dentry_iput,
+	.d_delete		= ocfs2_dentry_delete,
 };
diff --git a/fs/ocfs2/dir.c b/fs/ocfs2/dir.c
index bdfba9db558a..4bbe044c0bc1 100644
--- a/fs/ocfs2/dir.c
+++ b/fs/ocfs2/dir.c
@@ -29,6 +29,7 @@
 #include <linux/quotaops.h>
 #include <linux/sort.h>
 #include <linux/iversion.h>
+#include <linux/delay.h>
 
 #include <cluster/masklog.h>
 
@@ -37,7 +38,7 @@
 #include "alloc.h"
 #include "blockcheck.h"
 #include "dir.h"
-#include "dlmglue.h"
+#include "lockglue.h"
 #include "extent_map.h"
 #include "file.h"
 #include "inode.h"
@@ -54,6 +55,7 @@
 #define NAMEI_RA_CHUNKS  2
 #define NAMEI_RA_BLOCKS  4
 #define NAMEI_RA_SIZE        (NAMEI_RA_CHUNKS * NAMEI_RA_BLOCKS)
+#define NAMEI_RA_INDEX(c, b)  (((c) * NAMEI_RA_BLOCKS) + (b))
 
 static int ocfs2_do_extend_dir(struct super_block *sb,
 			       handle_t *handle,
@@ -315,8 +317,9 @@ static int ocfs2_check_dir_entry(struct inode * dir,
 		error_msg = "directory entry across blocks";
 
 	if (unlikely(error_msg != NULL))
-		mlog(ML_ERROR, "bad entry in directory #%llu: %s - "
+		mlog(ML_ERROR, "%s: bad entry in directory #%llu: %s - "
 		     "offset=%lu, inode=%llu, rec_len=%d, name_len=%d\n",
+		     OCFS2_SB(dir->i_sb)->uuid_str,
 		     (unsigned long long)OCFS2_I(dir)->ip_blkno, error_msg,
 		     offset, (unsigned long long)le64_to_cpu(de->inode), rlen,
 		     de->name_len);
@@ -444,8 +447,8 @@ static int ocfs2_validate_dir_block(struct super_block *sb,
 	 */
 	rc = ocfs2_validate_meta_ecc(sb, bh->b_data, &trailer->db_check);
 	if (rc)
-		mlog(ML_ERROR, "Checksum failed for dinode %llu\n",
-		     (unsigned long long)bh->b_blocknr);
+		mlog(ML_ERROR, "%s: Checksum failed for dinode %llu\n",
+		     OCFS2_SB(sb)->uuid_str, (unsigned long long)bh->b_blocknr);
 
 	return rc;
 }
@@ -463,26 +466,29 @@ static int ocfs2_check_dir_trailer(struct inode *dir, struct buffer_head *bh)
 
 	trailer = ocfs2_trailer_from_bh(bh, dir->i_sb);
 	if (!OCFS2_IS_VALID_DIR_TRAILER(trailer)) {
-		rc = ocfs2_error(dir->i_sb,
-				 "Invalid dirblock #%llu: signature = %.*s\n",
-				 (unsigned long long)bh->b_blocknr, 7,
-				 trailer->db_signature);
+		rc = -EINVAL;
+		ocfs2_error(dir->i_sb,
+			    "Invalid dirblock #%llu: signature = %.*s",
+			    (unsigned long long)bh->b_blocknr, 7,
+			    trailer->db_signature);
 		goto out;
 	}
 	if (le64_to_cpu(trailer->db_blkno) != bh->b_blocknr) {
-		rc = ocfs2_error(dir->i_sb,
-				 "Directory block #%llu has an invalid db_blkno of %llu\n",
-				 (unsigned long long)bh->b_blocknr,
-				 (unsigned long long)le64_to_cpu(trailer->db_blkno));
+		rc = -EINVAL;
+		ocfs2_error(dir->i_sb,
+			    "Directory block #%llu has an invalid db_blkno of %llu",
+			    (unsigned long long)bh->b_blocknr,
+			    (unsigned long long)le64_to_cpu(trailer->db_blkno));
 		goto out;
 	}
 	if (le64_to_cpu(trailer->db_parent_dinode) !=
 	    OCFS2_I(dir)->ip_blkno) {
-		rc = ocfs2_error(dir->i_sb,
-				 "Directory block #%llu on dinode #%llu has an invalid parent_dinode of %llu\n",
-				 (unsigned long long)bh->b_blocknr,
-				 (unsigned long long)OCFS2_I(dir)->ip_blkno,
-				 (unsigned long long)le64_to_cpu(trailer->db_blkno));
+		rc = -EINVAL;
+		ocfs2_error(dir->i_sb,
+			    "Directory block #%llu on dinode #%llu has an invalid parent_dinode of %llu",
+			    (unsigned long long)bh->b_blocknr,
+			    (unsigned long long)OCFS2_I(dir)->ip_blkno,
+			    (unsigned long long)le64_to_cpu(trailer->db_blkno));
 		goto out;
 	}
 out:
@@ -574,19 +580,20 @@ static int ocfs2_validate_dx_root(struct super_block *sb,
 	ret = ocfs2_validate_meta_ecc(sb, bh->b_data, &dx_root->dr_check);
 	if (ret) {
 		mlog(ML_ERROR,
-		     "Checksum failed for dir index root block %llu\n",
-		     (unsigned long long)bh->b_blocknr);
+		     "%s: Checksum failed for dir index root block %llu\n",
+		     OCFS2_SB(sb)->uuid_str, (unsigned long long)bh->b_blocknr);
 		return ret;
 	}
 
 	if (!OCFS2_IS_VALID_DX_ROOT(dx_root)) {
-		ret = ocfs2_error(sb,
-				  "Dir Index Root # %llu has bad signature %.*s\n",
-				  (unsigned long long)le64_to_cpu(dx_root->dr_blkno),
-				  7, dx_root->dr_signature);
+		ocfs2_error(sb,
+			    "Dir Index Root # %llu has bad signature %.*s",
+			    (unsigned long long)le64_to_cpu(dx_root->dr_blkno),
+			    7, dx_root->dr_signature);
+		return -EINVAL;
 	}
 
-	return ret;
+	return 0;
 }
 
 static int ocfs2_read_dx_root(struct inode *dir, struct ocfs2_dinode *di,
@@ -617,17 +624,18 @@ static int ocfs2_validate_dx_leaf(struct super_block *sb,
 	ret = ocfs2_validate_meta_ecc(sb, bh->b_data, &dx_leaf->dl_check);
 	if (ret) {
 		mlog(ML_ERROR,
-		     "Checksum failed for dir index leaf block %llu\n",
-		     (unsigned long long)bh->b_blocknr);
+		     "%s: Checksum failed for dir index leaf block %llu\n",
+		     OCFS2_SB(sb)->uuid_str, (unsigned long long)bh->b_blocknr);
 		return ret;
 	}
 
 	if (!OCFS2_IS_VALID_DX_LEAF(dx_leaf)) {
-		ret = ocfs2_error(sb, "Dir Index Leaf has bad signature %.*s\n",
-				  7, dx_leaf->dl_signature);
+		ocfs2_error(sb, "Dir Index Leaf has bad signature %.*s",
+			    7, dx_leaf->dl_signature);
+		return -EROFS;
 	}
 
-	return ret;
+	return 0;
 }
 
 static int ocfs2_read_dx_leaf(struct inode *dir, u64 blkno,
@@ -676,7 +684,7 @@ static struct buffer_head *ocfs2_find_entry_el(const char *name, int namelen,
 	int ra_ptr = 0;		/* Current index into readahead
 				   buffer */
 	int num = 0;
-	int nblocks, i;
+	int nblocks, i, err;
 
 	sb = dir->i_sb;
 
@@ -708,7 +716,7 @@ static struct buffer_head *ocfs2_find_entry_el(const char *name, int namelen,
 				num++;
 
 				bh = NULL;
-				ocfs2_read_dir_block(dir, b++, &bh,
+				err = ocfs2_read_dir_block(dir, b++, &bh,
 							   OCFS2_BH_READAHEAD);
 				bh_use[ra_max] = bh;
 			}
@@ -786,10 +794,10 @@ static int ocfs2_dx_dir_lookup_rec(struct inode *inode,
 		el = &eb->h_list;
 
 		if (el->l_tree_depth) {
-			ret = ocfs2_error(inode->i_sb,
-					  "Inode %lu has non zero tree depth in btree tree block %llu\n",
-					  inode->i_ino,
-					  (unsigned long long)eb_bh->b_blocknr);
+			ocfs2_error(inode->i_sb,
+				    "Inode %lu has non zero tree depth in btree tree block %llu", inode->i_ino,
+				    (unsigned long long)eb_bh->b_blocknr);
+			ret = -EROFS;
 			goto out;
 		}
 	}
@@ -805,11 +813,12 @@ static int ocfs2_dx_dir_lookup_rec(struct inode *inode,
 	}
 
 	if (!found) {
-		ret = ocfs2_error(inode->i_sb,
-				  "Inode %lu has bad extent record (%u, %u, 0) in btree\n",
-				  inode->i_ino,
-				  le32_to_cpu(rec->e_cpos),
-				  ocfs2_rec_clusters(el, rec));
+		ocfs2_error(inode->i_sb,
+				"Inode %lu has bad extent record (%u, %u, 0) in btree",
+				inode->i_ino,
+				le32_to_cpu(rec->e_cpos),
+				ocfs2_rec_clusters(el, rec));
+		ret = -EROFS;
 		goto out;
 	}
 
@@ -848,9 +857,9 @@ static int ocfs2_dx_dir_lookup(struct inode *inode,
 			       u64 *ret_phys_blkno)
 {
 	int ret = 0;
-	unsigned int cend, clen;
-	u32 cpos;
-	u64 blkno;
+	unsigned int cend, clen = 0;
+	u32 cpos = 0;
+	u64 blkno = 0UL;
 	u32 name_hash = hinfo->major_hash;
 
 	ret = ocfs2_dx_dir_lookup_rec(inode, el, name_hash, &cpos, &blkno,
@@ -894,7 +903,7 @@ static int ocfs2_dx_dir_search(const char *name, int namelen,
 			       struct ocfs2_dir_lookup_result *res)
 {
 	int ret, i, found;
-	u64 phys;
+	u64 phys = 0UL;
 	struct buffer_head *dx_leaf_bh = NULL;
 	struct ocfs2_dx_leaf *dx_leaf;
 	struct ocfs2_dx_entry *dx_entry = NULL;
@@ -1268,7 +1277,8 @@ static int ocfs2_delete_entry_dx(handle_t *handle, struct inode *dir,
 	index /= sizeof(*dx_entry);
 
 	if (index >= le16_to_cpu(entry_list->de_num_used)) {
-		mlog(ML_ERROR, "Dir %llu: Bad dx_entry ptr idx %d, (%p, %p)\n",
+		mlog(ML_ERROR, "%s: Dir %llu: Bad dx_entry ptr idx %d, (%p, %p)\n",
+		     OCFS2_SB(dir->i_sb)->uuid_str,
 		     (unsigned long long)OCFS2_I(dir)->ip_blkno, index,
 		     entry_list, dx_entry);
 		return -EIO;
@@ -1743,10 +1753,19 @@ static int ocfs2_dir_foreach_blk_id(struct inode *inode,
 	struct ocfs2_inline_data *data;
 	struct ocfs2_dir_entry *de;
 
+	if (ocfs2_test_invalid_fs(OCFS2_SB(inode->i_sb))) {
+		mlog(ML_ERROR,
+				"%s: OCFS2 invalid because of disk timeout.\n",
+				OCFS2_SB(inode->i_sb)->uuid_str);
+		return -EIO;
+	}
+
 	ret = ocfs2_read_inode_block(inode, &di_bh);
 	if (ret) {
-		mlog(ML_ERROR, "Unable to read inode block for dir %llu\n",
+		mlog(ML_ERROR, "%s: Unable to read inode block for dir %llu\n",
+		     OCFS2_SB(inode->i_sb)->uuid_str,
 		     (unsigned long long)OCFS2_I(inode)->ip_blkno);
+		msleep(1000);
 		goto out;
 	}
 
@@ -1931,6 +1950,12 @@ int ocfs2_readdir(struct file *file, struct dir_context *ctx)
 
 	trace_ocfs2_readdir((unsigned long long)OCFS2_I(inode)->ip_blkno);
 
+	if (ocfs2_test_invalid_fs(OCFS2_SB(inode->i_sb))) {
+		printk_once(KERN_ERR "%s: %s: OCFS2 invalid.\n",
+				__func__, OCFS2_SB(inode->i_sb)->uuid_str);
+		return -EIO;
+	}
+
 	error = ocfs2_inode_lock_atime(inode, file->f_path.mnt, &lock_level, 1);
 	if (lock_level && error >= 0) {
 		/* We release EX lock which used to update atime
@@ -2134,7 +2159,8 @@ int ocfs2_empty_dir(struct inode *inode)
 		mlog_errno(ret);
 
 	if (!priv.seen_dot || !priv.seen_dot_dot) {
-		mlog(ML_ERROR, "bad directory (dir #%llu) - no `.' or `..'\n",
+		mlog(ML_ERROR, "%s: bad directory (dir #%llu) - no `.' or `..'\n",
+		     OCFS2_SB(inode->i_sb)->uuid_str,
 		     (unsigned long long)OCFS2_I(inode)->ip_blkno);
 		/*
 		 * XXX: Is it really safe to allow an unlink to continue?
@@ -3718,8 +3744,8 @@ static int ocfs2_dx_dir_rebalance(struct ocfs2_super *osb, struct inode *dir,
 
 	num_used = le16_to_cpu(dx_leaf->dl_list.de_num_used);
 	if (num_used < le16_to_cpu(dx_leaf->dl_list.de_count)) {
-		mlog(ML_ERROR, "DX Dir: %llu, Asked to rebalance empty leaf: "
-		     "%llu, %d\n", (unsigned long long)OCFS2_I(dir)->ip_blkno,
+		mlog(ML_ERROR, "%s: DX Dir: %llu, Asked to rebalance empty leaf: %llu, %d\n",
+		     osb->uuid_str, (unsigned long long)OCFS2_I(dir)->ip_blkno,
 		     (unsigned long long)leaf_blkno, num_used);
 		ret = -EIO;
 		goto out;
@@ -4393,9 +4419,9 @@ static int ocfs2_dx_dir_remove_index(struct inode *dir,
 int ocfs2_dx_dir_truncate(struct inode *dir, struct buffer_head *di_bh)
 {
 	int ret;
-	unsigned int clen;
-	u32 major_hash = UINT_MAX, p_cpos, cpos;
-	u64 blkno;
+	unsigned int clen = 0;
+	u32 major_hash = UINT_MAX, p_cpos, cpos = 0;
+	u64 blkno = 0UL;
 	struct ocfs2_super *osb = OCFS2_SB(dir->i_sb);
 	struct buffer_head *dx_root_bh = NULL;
 	struct ocfs2_dx_root_block *dx_root;
diff --git a/fs/ocfs2/export.c b/fs/ocfs2/export.c
index 69ed278dd84d..cd4f5236883a 100644
--- a/fs/ocfs2/export.c
+++ b/fs/ocfs2/export.c
@@ -18,7 +18,7 @@
 
 #include "alloc.h"
 #include "dir.h"
-#include "dlmglue.h"
+#include "lockglue.h"
 #include "dcache.h"
 #include "export.h"
 #include "inode.h"
@@ -63,7 +63,7 @@ static struct dentry *ocfs2_get_dentry(struct super_block *sb,
 	 */
 	status = ocfs2_nfs_sync_lock(osb, 1);
 	if (status < 0) {
-		mlog(ML_ERROR, "getting nfs sync lock(EX) failed %d\n", status);
+		mlog(ML_ERROR, "%s: getting nfs sync lock(EX) failed %d\n", osb->uuid_str, status);
 		goto check_err;
 	}
 
@@ -77,7 +77,7 @@ static struct dentry *ocfs2_get_dentry(struct super_block *sb,
 			 */
 			status = -ESTALE;
 		} else
-			mlog(ML_ERROR, "test inode bit failed %d\n", status);
+			mlog(ML_ERROR, "%s: test inode bit failed %d\n", osb->uuid_str, status);
 		goto unlock_nfs_sync;
 	}
 
@@ -143,7 +143,7 @@ static struct dentry *ocfs2_get_parent(struct dentry *child)
 	if (status < 0) {
 		mlog(ML_ERROR, "getting nfs sync lock(EX) failed %d\n", status);
 		parent = ERR_PTR(status);
-		goto bail;
+		goto unlock_nfs_sync;
 	}
 
 	status = ocfs2_inode_lock(dir, NULL, 0);
@@ -151,7 +151,7 @@ static struct dentry *ocfs2_get_parent(struct dentry *child)
 		if (status != -ENOENT)
 			mlog_errno(status);
 		parent = ERR_PTR(status);
-		goto unlock_nfs_sync;
+		goto bail;
 	}
 
 	status = ocfs2_lookup_ino_from_name(dir, "..", 2, &blkno);
diff --git a/fs/ocfs2/extent_map.c b/fs/ocfs2/extent_map.c
index 7b93e9c766f6..48822cb03613 100644
--- a/fs/ocfs2/extent_map.c
+++ b/fs/ocfs2/extent_map.c
@@ -14,14 +14,16 @@
 #include <linux/slab.h>
 #include <linux/types.h>
 #include <linux/fiemap.h>
-
+#include <linux/blkdev.h>
 #include <cluster/masklog.h>
 
 #include "ocfs2.h"
 
 #include "alloc.h"
-#include "dlmglue.h"
+#include "lockglue.h"
 #include "extent_map.h"
+#include "file.h"
+#include "journal.h"
 #include "inode.h"
 #include "super.h"
 #include "symlink.h"
@@ -395,7 +397,7 @@ int ocfs2_figure_hole_clusters(struct ocfs2_caching_info *ci,
 	return ret;
 }
 
-static int ocfs2_get_clusters_nocache(struct inode *inode,
+int ocfs2_get_clusters_nocache(struct inode *inode,
 				      struct buffer_head *di_bh,
 				      u32 v_cluster, unsigned int *hole_len,
 				      struct ocfs2_extent_rec *ret_rec,
@@ -403,7 +405,7 @@ static int ocfs2_get_clusters_nocache(struct inode *inode,
 {
 	int i, ret, tree_height, len;
 	struct ocfs2_dinode *di;
-	struct ocfs2_extent_block *eb;
+	struct ocfs2_extent_block *eb = NULL;
 	struct ocfs2_extent_list *el;
 	struct ocfs2_extent_rec *rec;
 	struct buffer_head *eb_bh = NULL;
@@ -463,10 +465,8 @@ static int ocfs2_get_clusters_nocache(struct inode *inode,
 	BUG_ON(v_cluster < le32_to_cpu(rec->e_cpos));
 
 	if (!rec->e_blkno) {
-		ocfs2_error(inode->i_sb,
-			    "Inode %lu has bad extent record (%u, %u, 0)\n",
-			    inode->i_ino,
-			    le32_to_cpu(rec->e_cpos),
+		ocfs2_error(inode->i_sb, "Inode %lu has bad extent record (%u, %u, 0)",
+			    inode->i_ino, le32_to_cpu(rec->e_cpos),
 			    ocfs2_rec_clusters(el, rec));
 		ret = -EROFS;
 		goto out;
@@ -599,7 +599,7 @@ int ocfs2_get_clusters(struct inode *inode, u32 v_cluster,
 		       unsigned int *extent_flags)
 {
 	int ret;
-	unsigned int hole_len, flags = 0;
+	unsigned int hole_len = 0, flags = 0;
 	struct buffer_head *di_bh = NULL;
 	struct ocfs2_extent_rec rec;
 
@@ -744,6 +744,12 @@ int ocfs2_fiemap(struct inode *inode, struct fiemap_extent_info *fieinfo,
 	struct buffer_head *di_bh = NULL;
 	struct ocfs2_extent_rec rec;
 
+	if (ocfs2_test_invalid_fs(osb)) {
+		mlog(ML_ERROR,
+			"%s: OCFS2 invalid because of disk timeout.\n", osb->uuid_str);
+		return -EIO;
+	}
+
 	ret = fiemap_prep(inode, fieinfo, map_start, &map_len, 0);
 	if (ret)
 		return ret;
@@ -987,7 +993,8 @@ int ocfs2_read_virt_blocks(struct inode *inode, u64 v_block, int nr,
 		if (!p_block) {
 			rc = -EIO;
 			mlog(ML_ERROR,
-			     "Inode #%llu contains a hole at offset %llu\n",
+			     "%s: Inode #%llu contains a hole at offset %llu\n",
+			     OCFS2_SB(inode->i_sb)->uuid_str,
 			     (unsigned long long)OCFS2_I(inode)->ip_blkno,
 			     (unsigned long long)(v_block + done) <<
 			     inode->i_sb->s_blocksize_bits);
diff --git a/fs/ocfs2/extent_map.h b/fs/ocfs2/extent_map.h
index e5464f6cee8a..23341e0455a2 100644
--- a/fs/ocfs2/extent_map.h
+++ b/fs/ocfs2/extent_map.h
@@ -34,6 +34,9 @@ void ocfs2_extent_map_insert_rec(struct inode *inode,
 
 int ocfs2_get_clusters(struct inode *inode, u32 v_cluster, u32 *p_cluster,
 		       u32 *num_clusters, unsigned int *extent_flags);
+int ocfs2_get_clusters_nocache(struct inode *inode, struct buffer_head *di_bh,
+		u32 v_cluster, unsigned int *hole_len, struct ocfs2_extent_rec *ret_rec,
+		unsigned int *is_last);
 int ocfs2_extent_map_get_blocks(struct inode *inode, u64 v_blkno, u64 *p_blkno,
 				u64 *ret_count, unsigned int *extent_flags);
 
diff --git a/fs/ocfs2/ioctl.c b/fs/ocfs2/ioctl.c
index 89984172fc4a..f9be72fbafea 100644
--- a/fs/ocfs2/ioctl.c
+++ b/fs/ocfs2/ioctl.c
@@ -15,7 +15,7 @@
 
 #include "ocfs2.h"
 #include "alloc.h"
-#include "dlmglue.h"
+#include "lockglue.h"
 #include "file.h"
 #include "inode.h"
 #include "journal.h"
@@ -29,6 +29,7 @@
 #include "buffer_head_io.h"
 #include "suballoc.h"
 #include "move_extents.h"
+#include "super.h"
 
 #define o2info_from_user(a, b)	\
 		copy_from_user(&(a), (b), sizeof(a))
@@ -270,6 +271,29 @@ static int ocfs2_info_handle_journal_size(struct inode *inode,
 	return 0;
 }
 
+static int ocfs2_info_handle_journal_feature(struct inode *inode,
+		struct ocfs2_info_request __user *req)
+{
+	struct ocfs2_info_journal_feature oij;
+	struct ocfs2_super *osb = OCFS2_SB(inode->i_sb);
+	journal_superblock_t *sb;
+
+	if (o2info_from_user(oij, req))
+		return -EFAULT;
+
+	sb = osb->journal->j_journal->j_superblock;
+	oij.opt_compat = be32_to_cpu(sb->s_feature_compat);
+	oij.opt_incompat = be32_to_cpu(sb->s_feature_incompat);
+	oij.opt_ro_compat = be32_to_cpu(sb->s_feature_ro_compat);
+
+	o2info_set_request_filled(&oij.ij_req);
+
+	if (o2info_to_user(oij, req))
+		return -EFAULT;
+
+	return 0;
+}
+
 static int ocfs2_info_scan_inode_alloc(struct ocfs2_super *osb,
 				       struct inode *inode_alloc, u64 blkno,
 				       struct ocfs2_info_freeinode *fi,
@@ -347,8 +371,8 @@ static int ocfs2_info_handle_freeinode(struct inode *inode,
 		if (o2info_coherent(&oifi->ifi_req)) {
 			inode_alloc = ocfs2_get_system_file_inode(osb, type, i);
 			if (!inode_alloc) {
-				mlog(ML_ERROR, "unable to get alloc inode in "
-				    "slot %u\n", i);
+				mlog(ML_ERROR, "%s: unable to get alloc inode in slot %u\n",
+					osb->uuid_str, i);
 				status = -EIO;
 				goto bail;
 			}
@@ -464,8 +488,8 @@ static int ocfs2_info_freefrag_scan_chain(struct ocfs2_super *osb,
 			status = ocfs2_read_blocks_sync(osb, blkno, 1, &bh);
 
 		if (status < 0) {
-			mlog(ML_ERROR, "Can't read the group descriptor # "
-			     "%llu from device.", (unsigned long long)blkno);
+			mlog(ML_ERROR, "%s: Can't read the group descriptor # %llu from device.\n",
+				 osb->uuid_str, (unsigned long long)blkno);
 			status = -EIO;
 			goto bail;
 		}
@@ -601,6 +625,7 @@ static int ocfs2_info_freefrag_scan_bitmap(struct ocfs2_super *osb,
 		inode_unlock(gb_inode);
 
 	iput(gb_inode);
+
 	brelse(bh);
 
 	return status;
@@ -641,7 +666,8 @@ static int ocfs2_info_handle_freefrag(struct inode *inode,
 		gb_inode = ocfs2_get_system_file_inode(osb, type,
 						       OCFS2_INVALID_SLOT);
 		if (!gb_inode) {
-			mlog(ML_ERROR, "unable to get global_bitmap inode\n");
+			mlog(ML_ERROR, "%s: unable to get global_bitmap inode\n",
+					osb->uuid_str);
 			status = -EIO;
 			goto bail;
 		}
@@ -744,6 +770,10 @@ static int ocfs2_info_handle_request(struct inode *inode,
 		if (oir.ir_size == sizeof(struct ocfs2_info_journal_size))
 			status = ocfs2_info_handle_journal_size(inode, req);
 		break;
+	case OCFS2_INFO_JOURNAL_FEATURE:
+		if (oir.ir_size == sizeof(struct ocfs2_info_journal_feature))
+			status = ocfs2_info_handle_journal_feature(inode, req);
+		break;
 	case OCFS2_INFO_FREEINODE:
 		if (oir.ir_size == sizeof(struct ocfs2_info_freeinode))
 			status = ocfs2_info_handle_freeinode(inode, req);
@@ -833,13 +863,69 @@ static int ocfs2_info_handle(struct inode *inode, struct ocfs2_info *info,
 	return status;
 }
 
+static int ocfs2_get_ws_state(struct inode *inode, unsigned int *flags)
+{
+	struct ocfs2_super *osb = OCFS2_SB(inode->i_sb);
+	int ws_switch;
+
+	ws_switch = ocfs2_get_vaai_switch_state(osb, OCFS2_WS_STATE);
+	if (ws_switch < 0) {
+		mlog_errno(ws_switch);
+		return ws_switch;
+	}
+
+	spin_lock(&osb->osb_lock);
+	if (osb->osb_ws_support && ws_switch)
+		*flags = 1;
+	else
+		*flags = 0;
+	spin_unlock(&osb->osb_lock);
+	return 0;
+}
+
+static int ocfs2_get_xcopy_state(struct inode *inode, unsigned int *flags)
+{
+	struct ocfs2_super *osb = OCFS2_SB(inode->i_sb);
+	int xcopy_switch;
+
+	xcopy_switch = ocfs2_get_vaai_switch_state(osb, OCFS2_XCOPY_STATE);
+	if (xcopy_switch < 0) {
+		mlog_errno(xcopy_switch);
+		return xcopy_switch;
+	}
+
+	spin_lock(&osb->osb_lock);
+	if (osb->osb_xcopy_support && xcopy_switch)
+		*flags = 1;
+	else
+		*flags = 0;
+	spin_unlock(&osb->osb_lock);
+
+	return 0;
+}
+
+static void ocfs2_get_fs_status(struct inode *inode, unsigned int *flags)
+{
+	struct ocfs2_super *osb = OCFS2_SB(inode->i_sb);
+
+#define OCFS2_STATUS_OTHER 0
+#define OCFS2_STATUS_INVALID 1
+#define OCFS2_STATUS_REPAIR 2
+	if (ocfs2_test_invalid_fs(osb))
+		*flags = OCFS2_STATUS_INVALID;
+	else if (ocfs2_test_repairing_fs(osb))
+		*flags = OCFS2_STATUS_REPAIR;
+	else
+		*flags = OCFS2_STATUS_OTHER;
+}
+
 long ocfs2_ioctl(struct file *filp, unsigned int cmd, unsigned long arg)
 {
 	struct inode *inode = file_inode(filp);
 	unsigned int flags;
-	int new_clusters;
 	int status;
 	struct ocfs2_space_resv sr;
+	struct ocfs2_extend_group_input support_input;
 	struct ocfs2_new_group_input input;
 	struct reflink_arguments args;
 	const char __user *old_path;
@@ -847,6 +933,11 @@ long ocfs2_ioctl(struct file *filp, unsigned int cmd, unsigned long arg)
 	bool preserve;
 	struct ocfs2_info info;
 	void __user *argp = (void __user *)arg;
+	struct ocfs2_zero_range ocfs2_range;
+	struct ocfs2_device_extent_list dev_extent_list;
+	struct ocfs2_device_extent_list __user *udev_extent_list =
+			(struct ocfs2_device_extent_list __user *) arg;
+	struct ocfs2_device_extent_info dev_extent_info = { 0, };
 
 	switch (cmd) {
 	case OCFS2_IOC_GETFLAGS:
@@ -879,13 +970,15 @@ long ocfs2_ioctl(struct file *filp, unsigned int cmd, unsigned long arg)
 		if (!capable(CAP_SYS_RESOURCE))
 			return -EPERM;
 
-		if (get_user(new_clusters, (int __user *)arg))
+		if (copy_from_user(&support_input,
+				(struct ocfs2_extend_group_input __user *)arg,
+				sizeof(struct ocfs2_extend_group_input)))
 			return -EFAULT;
 
 		status = mnt_want_write_file(filp);
 		if (status)
 			return status;
-		status = ocfs2_group_extend(inode, new_clusters);
+		status = ocfs2_group_extend(inode, cmd, &support_input);
 		mnt_drop_write_file(filp);
 		return status;
 	case OCFS2_IOC_GROUP_ADD:
@@ -899,7 +992,7 @@ long ocfs2_ioctl(struct file *filp, unsigned int cmd, unsigned long arg)
 		status = mnt_want_write_file(filp);
 		if (status)
 			return status;
-		status = ocfs2_group_add(inode, &input);
+		status = ocfs2_group_add(inode, cmd, &input);
 		mnt_drop_write_file(filp);
 		return status;
 	case OCFS2_IOC_REFLINK:
@@ -944,6 +1037,70 @@ long ocfs2_ioctl(struct file *filp, unsigned int cmd, unsigned long arg)
 	}
 	case OCFS2_IOC_MOVE_EXT:
 		return ocfs2_ioctl_move_extents(filp, argp);
+	case OCFS2_ZERO_UPDATE:
+	case OCFS2_IOC_ZERO:
+		if (copy_from_user(&ocfs2_range, (struct ocfs2_zero_range __user *)arg,
+				sizeof(ocfs2_range)))
+			return -EFAULT;
+		return ocfs2_zero_file(filp, cmd, &ocfs2_range);
+	case OCFS2_IOC_XCOPY_READ:
+	case OCFS2_IOC_DEVADDR_QUERY:
+		if (copy_from_user(&dev_extent_list, udev_extent_list,
+				sizeof(dev_extent_list)))
+			return -EFAULT;
+
+		if (dev_extent_list.max_extent_count > OCFS2_XCOPY_MAX_EXTENTS)
+			return -EINVAL;
+
+		dev_extent_info.ei_extent_count = dev_extent_list.max_extent_count;
+		dev_extent_info.ei_extents_start = udev_extent_list->extent;
+
+		status = ocfs2_query_device_addr(filp, dev_extent_list.start,
+				dev_extent_list.len, &dev_extent_info, cmd);
+		if (status)
+			return status;
+
+		dev_extent_list.mapped_extents = dev_extent_info.ei_mapped_extents;
+		if (copy_to_user(udev_extent_list, &dev_extent_list,
+				sizeof(dev_extent_list)))
+			status = -EFAULT;
+
+		return status;
+	case OCFS2_IOC_XCOPY_WRITE:
+		if (copy_from_user(&dev_extent_list, udev_extent_list,
+				sizeof(dev_extent_list)))
+			return -EFAULT;
+
+		if (dev_extent_list.max_extent_count > OCFS2_XCOPY_MAX_EXTENTS)
+			return -EINVAL;
+
+		dev_extent_info.ei_mapped_extents = dev_extent_list.mapped_extents;
+		dev_extent_info.ei_extents_start = udev_extent_list->extent;
+
+		return ocfs2_extended_copy(filp, dev_extent_list.start,
+				dev_extent_list.len, &dev_extent_info);
+	case OCFS2_IOC_GET_XCOPY_FLAGS:
+		if (ocfs2_get_xcopy_state(inode, &flags) < 0)
+			return -EINVAL;
+		return put_user(flags, (int __user *) arg);
+	case OCFS2_IOC_GET_WS_FLAGS:
+		if (ocfs2_get_ws_state(inode, &flags) < 0)
+			return -EINVAL;
+		return put_user(flags, (int __user *)arg);
+	case OCFS2_IOC_UPDATE_STOR_ATTR:
+	case OCFS2_IOC_SET_HIGH_PERF_DIO:
+		if (get_user(flags, (int __user *) arg))
+			return -EFAULT;
+
+		return ocfs2_update_super_and_backups(inode, cmd, &flags);
+	case OCFS2_IOC_FENCE_SCSI:
+	case OCFS2_IOC_CLEAR_FENCE_SCSI:
+		return ocfs2_update_super_and_backups(inode, cmd, NULL);
+	case OCFS2_IOC_GET_FS_STATUS:
+		ocfs2_get_fs_status(inode, &flags);
+		return put_user(flags, (int __user *)arg);
+	case OCFS2_IOC_RESORT_GD:
+		return ocfs2_resort_global_bitmap(inode);
 	default:
 		return -ENOTTY;
 	}
@@ -972,6 +1129,7 @@ long ocfs2_compat_ioctl(struct file *file, unsigned cmd, unsigned long arg)
 	case OCFS2_IOC_GROUP_EXTEND:
 	case OCFS2_IOC_GROUP_ADD:
 	case OCFS2_IOC_GROUP_ADD64:
+	case OCFS2_IOC_RESORT_GD:
 		break;
 	case OCFS2_IOC_REFLINK:
 		if (copy_from_user(&args, argp, sizeof(args)))
diff --git a/fs/ocfs2/mmap.c b/fs/ocfs2/mmap.c
index 25cabbfe87fc..cd7e593d8e75 100644
--- a/fs/ocfs2/mmap.c
+++ b/fs/ocfs2/mmap.c
@@ -22,7 +22,7 @@
 #include "ocfs2.h"
 
 #include "aops.h"
-#include "dlmglue.h"
+#include "lockglue.h"
 #include "file.h"
 #include "inode.h"
 #include "mmap.h"
@@ -121,6 +121,13 @@ static vm_fault_t ocfs2_page_mkwrite(struct vm_fault *vmf)
 	int err;
 	vm_fault_t ret;
 
+	if (ocfs2_test_invalid_fs(OCFS2_SB(inode->i_sb))) {
+		mlog(ML_ERROR,
+				"%s: OCFS2 invalid because of disk timeout.\n",
+				OCFS2_SB(inode->i_sb)->uuid_str);
+		return VM_FAULT_SIGBUS;
+	}
+
 	sb_start_pagefault(inode->i_sb);
 	ocfs2_block_signals(&oldset);
 
diff --git a/fs/ocfs2/move_extents.c b/fs/ocfs2/move_extents.c
index 758d9661ef1e..9dac85c7ab8c 100644
--- a/fs/ocfs2/move_extents.c
+++ b/fs/ocfs2/move_extents.c
@@ -19,7 +19,7 @@
 #include "alloc.h"
 #include "localalloc.h"
 #include "aops.h"
-#include "dlmglue.h"
+#include "lockglue.h"
 #include "extent_map.h"
 #include "inode.h"
 #include "journal.h"
@@ -92,9 +92,10 @@ static int __ocfs2_move_extent(handle_t *handle,
 
 	index = ocfs2_search_extent_list(el, cpos);
 	if (index == -1) {
-		ret = ocfs2_error(inode->i_sb,
-				  "Inode %llu has an extent at cpos %u which can no longer be found\n",
-				  (unsigned long long)ino, cpos);
+		ocfs2_error(inode->i_sb,
+			    "Inode %llu has an extent at cpos %u which can no longer be found.",
+			    (unsigned long long)ino, cpos);
+		ret = -EROFS;
 		goto out;
 	}
 
@@ -181,7 +182,6 @@ static int ocfs2_lock_meta_allocator_move_extents(struct inode *inode,
 		goto out;
 	}
 
-
 	*credits += ocfs2_calc_extend_credits(osb->sb, et->et_root_el);
 
 	mlog(0, "reserve metadata_blocks: %d, data_clusters: %u, credits: %d\n",
@@ -635,7 +635,7 @@ static int ocfs2_move_extent(struct ocfs2_move_extents_context *context,
 	gb_inode = ocfs2_get_system_file_inode(osb, GLOBAL_BITMAP_SYSTEM_INODE,
 					       OCFS2_INVALID_SLOT);
 	if (!gb_inode) {
-		mlog(ML_ERROR, "unable to get global_bitmap inode\n");
+		mlog(ML_ERROR, "%s: unable to get global_bitmap inode\n", osb->uuid_str);
 		ret = -EIO;
 		goto out;
 	}
@@ -908,7 +908,9 @@ static int ocfs2_move_extents(struct ocfs2_move_extents_context *context)
 	struct inode *inode = context->inode;
 	struct ocfs2_dinode *di;
 	struct buffer_head *di_bh = NULL;
-	struct ocfs2_super *osb = OCFS2_SB(inode->i_sb);
+	struct ocfs2_super *osb;
+
+	osb = OCFS2_SB(inode->i_sb);
 
 	if (ocfs2_is_hard_readonly(osb) || ocfs2_is_soft_readonly(osb))
 		return -EROFS;
diff --git a/fs/ocfs2/namei.c b/fs/ocfs2/namei.c
index 856474b0a1ae..3bc8839d2b24 100644
--- a/fs/ocfs2/namei.c
+++ b/fs/ocfs2/namei.c
@@ -29,6 +29,7 @@
 #include <linux/quotaops.h>
 #include <linux/iversion.h>
 
+#define MLOG_MASK_PREFIX ML_FILE_IO
 #include <cluster/masklog.h>
 
 #include "ocfs2.h"
@@ -36,7 +37,7 @@
 #include "alloc.h"
 #include "dcache.h"
 #include "dir.h"
-#include "dlmglue.h"
+#include "lockglue.h"
 #include "extent_map.h"
 #include "file.h"
 #include "inode.h"
@@ -53,6 +54,8 @@
 
 #include "buffer_head_io.h"
 
+#include "adl/adlapi.h"
+
 static int ocfs2_mknod_locked(struct ocfs2_super *osb,
 			      struct inode *dir,
 			      struct inode *inode,
@@ -60,7 +63,9 @@ static int ocfs2_mknod_locked(struct ocfs2_super *osb,
 			      struct buffer_head **new_fe_bh,
 			      struct buffer_head *parent_fe_bh,
 			      handle_t *handle,
-			      struct ocfs2_alloc_context *inode_ac);
+			      struct ocfs2_alloc_context *inode_ac,
+			      struct ocfs2_alloc_context *lock_ac,
+			      bool disk_lock);
 
 static int ocfs2_prepare_orphan_dir(struct ocfs2_super *osb,
 				    struct inode **ret_orphan_dir,
@@ -91,9 +96,45 @@ static int ocfs2_double_lock(struct ocfs2_super *osb,
 			     int rename);
 
 static void ocfs2_double_unlock(struct inode *inode1, struct inode *inode2);
+
 /* An orphan dir name is an 8 byte value, printed as a hex string */
 #define OCFS2_ORPHAN_NAMELEN ((int)(2 * sizeof(u64)))
 
+static int ocfs2_remove_entry(struct ocfs2_super *osb, struct inode *dir,
+		struct ocfs2_dir_lookup_result *lookup)
+{
+	handle_t *handle = NULL;
+	int status = 0;
+
+	/* get ex lock before delete entry from parent dir */
+	status = ocfs2_inode_lock(dir, NULL, 1);
+	if (status < 0) {
+		if (status != -ENOENT)
+			mlog_errno(status);
+		return status;
+	}
+
+	handle = ocfs2_start_trans(osb,
+			OCFS2_INODE_DEL_FROM_ORPHAN_CREDITS);
+	if (IS_ERR(handle)) {
+		status = PTR_ERR(handle);
+		mlog_errno(status);
+		goto unlock;
+	}
+
+	/* remove it from the parent directory */
+	status = ocfs2_delete_entry(handle, dir, lookup);
+	if (status < 0)
+		mlog_errno(status);
+
+	ocfs2_commit_trans(osb, handle);
+
+unlock:
+	ocfs2_inode_unlock(dir, 1);
+
+	return status;
+}
+
 static struct dentry *ocfs2_lookup(struct inode *dir, struct dentry *dentry,
 				   unsigned int flags)
 {
@@ -102,16 +143,29 @@ static struct dentry *ocfs2_lookup(struct inode *dir, struct dentry *dentry,
 	struct inode *inode = NULL;
 	struct dentry *ret;
 	struct ocfs2_inode_info *oi;
+	struct ocfs2_super *osb = OCFS2_SB(dir->i_sb);
+	struct ocfs2_dir_lookup_result lookup = { NULL, };
 
 	trace_ocfs2_lookup(dir, dentry, dentry->d_name.len,
 			   dentry->d_name.name,
 			   (unsigned long long)OCFS2_I(dir)->ip_blkno, 0);
 
+	/* do not allow to lookup in case of online recovery */
+	if (ocfs2_test_repairing_fs(osb)) {
+		ret = ERR_PTR(-EIO);
+		goto bail;
+	}
+
 	if (dentry->d_name.len > OCFS2_MAX_FILENAME_LEN) {
 		ret = ERR_PTR(-ENAMETOOLONG);
 		goto bail;
 	}
 
+	if (ocfs2_test_invalid_fs(osb)) {
+		ret = ERR_PTR(-EIO);
+		goto bail;
+	}
+
 	status = ocfs2_inode_lock_nested(dir, NULL, 0, OI_LS_PARENT);
 	if (status < 0) {
 		if (status != -ENOENT)
@@ -120,15 +174,30 @@ static struct dentry *ocfs2_lookup(struct inode *dir, struct dentry *dentry,
 		goto bail;
 	}
 
-	status = ocfs2_lookup_ino_from_name(dir, dentry->d_name.name,
-					    dentry->d_name.len, &blkno);
+	status = ocfs2_find_files_on_disk(dentry->d_name.name,
+			dentry->d_name.len, &blkno, dir, &lookup);
 	if (status < 0)
 		goto bail_add;
 
-	inode = ocfs2_iget(OCFS2_SB(dir->i_sb), blkno, 0, 0);
+	inode = ocfs2_iget(osb, blkno, OCFS2_FI_FLAG_LOOKUP, 0);
 	if (IS_ERR(inode)) {
-		ret = ERR_PTR(-EACCES);
-		goto bail_unlock;
+		if (PTR_ERR(inode) != -EROFS) {
+			ret = ERR_PTR(-EACCES);
+			mlog(ML_ERROR,
+					"%s: Lookup() iget failed. Name: %pd, blkno:%llu, dir: %llu\n",
+					osb->uuid_str, dentry, blkno,
+					(unsigned long long)OCFS2_I(dir)->ip_blkno);
+			goto bail_unlock;
+		} else {
+			ocfs2_inode_unlock(dir, 0);
+			mlog(ML_ERROR, "Invalid dinode %llu found in %pd, force delete it\n",
+					blkno, dentry);
+			status = ocfs2_remove_entry(osb, dir, &lookup);
+			if (status)
+				mlog_errno(status);
+			ret = ERR_PTR(-ENOENT);
+			goto bail;
+		}
 	}
 
 	oi = OCFS2_I(inode);
@@ -176,7 +245,7 @@ static struct dentry *ocfs2_lookup(struct inode *dir, struct dentry *dentry,
 	ocfs2_inode_unlock(dir, 0);
 
 bail:
-
+	ocfs2_free_dir_lookup_result(&lookup);
 	trace_ocfs2_lookup_ret(ret);
 
 	return ret;
@@ -189,7 +258,7 @@ static struct inode *ocfs2_get_init_inode(struct inode *dir, umode_t mode)
 
 	inode = new_inode(dir->i_sb);
 	if (!inode) {
-		mlog(ML_ERROR, "new_inode failed!\n");
+		mlog(ML_ERROR, "%s: new_inode failed!\n", OCFS2_SB(dir->i_sb)->uuid_str);
 		return ERR_PTR(-ENOMEM);
 	}
 
@@ -198,11 +267,12 @@ static struct inode *ocfs2_get_init_inode(struct inode *dir, umode_t mode)
 	 * callers. */
 	if (S_ISDIR(mode))
 		set_nlink(inode, 2);
+	else
+		set_nlink(inode, 1);
 	inode_init_owner(inode, dir, mode);
 	status = dquot_initialize(inode);
 	if (status)
 		return ERR_PTR(status);
-
 	return inode;
 }
 
@@ -210,13 +280,17 @@ static void ocfs2_cleanup_add_entry_failure(struct ocfs2_super *osb,
 		struct dentry *dentry, struct inode *inode)
 {
 	struct ocfs2_dentry_lock *dl = dentry->d_fsdata;
+	unsigned long flags;
+
+	if (ocfs2_disk_lock(osb))
+		return;
 
 	ocfs2_simple_drop_lockres(osb, &dl->dl_lockres);
-	ocfs2_lock_res_free(&dl->dl_lockres);
+	ocfs2_lock_res_free(osb, &dl->dl_lockres);
 	BUG_ON(dl->dl_count != 1);
-	spin_lock(&dentry_attach_lock);
+	spin_lock_irqsave(&dentry_attach_lock, flags);
 	dentry->d_fsdata = NULL;
-	spin_unlock(&dentry_attach_lock);
+	spin_unlock_irqrestore(&dentry_attach_lock, flags);
 	kfree(dl);
 	iput(inode);
 }
@@ -237,6 +311,7 @@ static int ocfs2_mknod(struct inode *dir,
 	struct ocfs2_alloc_context *inode_ac = NULL;
 	struct ocfs2_alloc_context *data_ac = NULL;
 	struct ocfs2_alloc_context *meta_ac = NULL;
+	struct ocfs2_alloc_context *lock_ac = NULL;
 	int want_clusters = 0;
 	int want_meta = 0;
 	int xattr_credits = 0;
@@ -248,6 +323,7 @@ static int ocfs2_mknod(struct inode *dir,
 	sigset_t oldset;
 	int did_block_signals = 0;
 	struct ocfs2_dentry_lock *dl = NULL;
+	bool disk_lock;
 
 	trace_ocfs2_mknod(dir, dentry, dentry->d_name.len, dentry->d_name.name,
 			  (unsigned long long)OCFS2_I(dir)->ip_blkno,
@@ -262,6 +338,12 @@ static int ocfs2_mknod(struct inode *dir,
 	/* get our super block */
 	osb = OCFS2_SB(dir->i_sb);
 
+	if (ocfs2_test_invalid_fs(osb)) {
+		printk_once(KERN_ERR "%s: %s: OCFS2 invalid.\n",
+				__func__, osb->uuid_str);
+		return -EIO;
+	}
+
 	status = ocfs2_inode_lock(dir, &parent_fe_bh, 1);
 	if (status < 0) {
 		if (status != -ENOENT)
@@ -295,6 +377,17 @@ static int ocfs2_mknod(struct inode *dir,
 		goto leave;
 	}
 
+	/* reserve an lock block */
+	disk_lock = ocfs2_disk_lock_support(osb);
+	if (disk_lock) {
+		status = ocfs2_reserve_new_lock(osb, &lock_ac);
+		if (status < 0) {
+			if (status != -ENOSPC)
+				mlog_errno(status);
+			goto leave;
+		}
+	}
+
 	/* reserve an inode spot */
 	status = ocfs2_reserve_new_inode(osb, &inode_ac);
 	if (status < 0) {
@@ -356,7 +449,7 @@ static int ocfs2_mknod(struct inode *dir,
 
 	handle = ocfs2_start_trans(osb, ocfs2_mknod_credits(osb->sb,
 							    S_ISDIR(mode),
-							    xattr_credits));
+							    xattr_credits, disk_lock));
 	if (IS_ERR(handle)) {
 		status = PTR_ERR(handle);
 		handle = NULL;
@@ -376,7 +469,7 @@ static int ocfs2_mknod(struct inode *dir,
 	/* do the real work now. */
 	status = ocfs2_mknod_locked(osb, dir, inode, dev,
 				    &new_fe_bh, parent_fe_bh, handle,
-				    inode_ac);
+				    inode_ac, lock_ac, disk_lock);
 	if (status < 0) {
 		mlog_errno(status);
 		goto leave;
@@ -402,10 +495,10 @@ static int ocfs2_mknod(struct inode *dir,
 		ocfs2_journal_dirty(handle, parent_fe_bh);
 		inc_nlink(dir);
 	}
-
 	status = ocfs2_init_acl(handle, inode, dir, new_fe_bh, parent_fe_bh,
 			 meta_ac, data_ac);
 
+
 	if (status < 0) {
 		mlog_errno(status);
 		goto roll_back;
@@ -438,10 +531,21 @@ static int ocfs2_mknod(struct inode *dir,
 				 OCFS2_I(inode)->ip_blkno, parent_fe_bh,
 				 &lookup);
 	if (status < 0) {
+		mlog(ML_ERROR,
+			"%s: ocfs2_add_entry failed, file %pd, dir %llu, nlink %u, inode %llu, nlink %u, status %d\n",
+			osb->uuid_str, dentry,
+			(unsigned long long)OCFS2_I(dir)->ip_blkno, dir->i_nlink,
+			(unsigned long long)OCFS2_I(inode)->ip_blkno, inode->i_nlink,
+			status);
 		mlog_errno(status);
 		goto roll_back;
 	}
 
+	mlog(0, "%s: create new file %pd success, dir %llu, nlink %u, inode %llu, nlink %u\n",
+			osb->uuid_str, dentry,
+			(unsigned long long)OCFS2_I(dir)->ip_blkno, dir->i_nlink,
+			(unsigned long long)OCFS2_I(inode)->ip_blkno, inode->i_nlink);
+
 	insert_inode_hash(inode);
 	d_instantiate(dentry, inode);
 	status = 0;
@@ -480,6 +584,9 @@ static int ocfs2_mknod(struct inode *dir,
 	if (meta_ac)
 		ocfs2_free_alloc_context(meta_ac);
 
+	if (lock_ac)
+		ocfs2_free_alloc_context(lock_ac);
+
 	/*
 	 * We should call iput after the i_mutex of the bitmap been
 	 * unlocked in ocfs2_free_alloc_context, or the
@@ -507,12 +614,13 @@ static int __ocfs2_mknod_locked(struct inode *dir,
 				struct buffer_head *parent_fe_bh,
 				handle_t *handle,
 				struct ocfs2_alloc_context *inode_ac,
+				struct ocfs2_alloc_context *lock_ac,
+				u64 fe_lock_blkno, u64 suballoc_lock_loc, u16 suballoc_lock_bit,
 				u64 fe_blkno, u64 suballoc_loc, u16 suballoc_bit)
 {
 	int status = 0;
 	struct ocfs2_super *osb = OCFS2_SB(dir->i_sb);
 	struct ocfs2_dinode *fe = NULL;
-	struct ocfs2_extent_list *fel;
 	u16 feat;
 	struct ocfs2_inode_info *oi = OCFS2_I(inode);
 	struct timespec64 ts;
@@ -524,10 +632,16 @@ static int __ocfs2_mknod_locked(struct inode *dir,
 	 * callers. */
 	inode->i_ino = ino_from_blkno(osb->sb, fe_blkno);
 	oi->ip_blkno = fe_blkno;
+	oi->ip_inode_lock_blkno = fe_lock_blkno;
 	spin_lock(&osb->osb_lock);
 	inode->i_generation = osb->s_next_generation++;
 	spin_unlock(&osb->osb_lock);
 
+	if (ocfs2_test_invalid_fs(osb)) {
+		status = -EIO;
+		goto leave;
+	}
+
 	*new_fe_bh = sb_getblk(osb->sb, fe_blkno);
 	if (!*new_fe_bh) {
 		status = -ENOMEM;
@@ -553,6 +667,11 @@ static int __ocfs2_mknod_locked(struct inode *dir,
 	fe->i_suballoc_loc = cpu_to_le64(suballoc_loc);
 	fe->i_suballoc_bit = cpu_to_le16(suballoc_bit);
 	fe->i_suballoc_slot = cpu_to_le16(inode_ac->ac_alloc_slot);
+	if (lock_ac)
+		fe->i_lock_suballoc_slot = lock_ac->ac_alloc_slot;
+	fe->i_lock_blkno = cpu_to_le64(fe_lock_blkno);
+	fe->i_lock_suballoc_bit = cpu_to_le16(suballoc_lock_bit);
+	fe->i_lock_suballoc_loc = cpu_to_le64(suballoc_lock_loc);
 	fe->i_uid = cpu_to_le32(i_uid_read(inode));
 	fe->i_gid = cpu_to_le32(i_gid_read(inode));
 	fe->i_mode = cpu_to_le16(inode->i_mode);
@@ -582,17 +701,26 @@ static int __ocfs2_mknod_locked(struct inode *dir,
 		fe->id2.i_data.id_count = cpu_to_le16(
 				ocfs2_max_inline_data_with_xattr(osb->sb, fe));
 	} else {
-		fel = &fe->id2.i_list;
-		fel->l_tree_depth = 0;
-		fel->l_next_free_rec = 0;
-		fel->l_count = cpu_to_le16(ocfs2_extent_recs_per_inode(osb->sb));
+		/*
+		 * Adjust extent record count to reserve
+		 * space for inline extended attribute.
+		 */
+		if (S_ISREG(inode->i_mode) &&
+				ocfs2_supports_xattr(osb) &&
+				!(osb->s_mount_opt & OCFS2_MOUNT_NOUSERXATTR)) {
+			fe->i_dyn_features = cpu_to_le16(feat | OCFS2_INLINE_XATTR_FL
+							      | OCFS2_HAS_XATTR_FL);
+			fe->i_xattr_inline_size = cpu_to_le16(osb->s_xattr_inline_size);
+		}
+		ocfs2_dinode_new_extent_list(inode, fe);
 	}
 
 	ocfs2_journal_dirty(handle, *new_fe_bh);
 
 	ocfs2_populate_inode(inode, fe, 1);
 	ocfs2_ci_set_new(osb, INODE_CACHE(inode));
-	if (!ocfs2_mount_local(osb)) {
+
+	if (!ocfs2_mount_local(osb) && !ocfs2_recover_ro(osb)) {
 		status = ocfs2_create_new_inode_locks(inode);
 		if (status < 0)
 			mlog_errno(status);
@@ -603,6 +731,8 @@ static int __ocfs2_mknod_locked(struct inode *dir,
 leave:
 	if (status < 0) {
 		if (*new_fe_bh) {
+			if (fe)
+				ocfs2_set_links_count(fe, 0);
 			brelse(*new_fe_bh);
 			*new_fe_bh = NULL;
 		}
@@ -620,25 +750,89 @@ static int ocfs2_mknod_locked(struct ocfs2_super *osb,
 			      struct buffer_head **new_fe_bh,
 			      struct buffer_head *parent_fe_bh,
 			      handle_t *handle,
-			      struct ocfs2_alloc_context *inode_ac)
+			      struct ocfs2_alloc_context *inode_ac,
+			      struct ocfs2_alloc_context *lock_ac,
+			      bool disk_lock)
 {
 	int status = 0;
 	u64 suballoc_loc, fe_blkno = 0;
 	u16 suballoc_bit;
+	u64 suballoc_lock_loc = 0, fe_lock_blkno = 0;
+	u16 suballoc_lock_bit = 0;
+	int lock_claimed = 0, inode_claimed = 0;
 
 	*new_fe_bh = NULL;
 
+	/* claim lock block */
+	if (disk_lock) {
+		status = ocfs2_claim_new_lock(handle,
+						lock_ac, &suballoc_lock_loc,
+						&suballoc_lock_bit, &fe_lock_blkno);
+		if (status < 0) {
+			mlog_errno(status);
+			return status;
+		}
+		lock_claimed = 1;
+
+		/* initialize disk block */
+		status = adl_initialize_lock_block(osb, fe_lock_blkno);
+		if (status) {
+			mlog_errno(status);
+			goto bail;
+		}
+	}
+
+	/* alloc inode block */
 	status = ocfs2_claim_new_inode(handle, dir, parent_fe_bh,
 				       inode_ac, &suballoc_loc,
 				       &suballoc_bit, &fe_blkno);
 	if (status < 0) {
 		mlog_errno(status);
-		return status;
+		goto bail;
 	}
+	inode_claimed = 1;
 
-	return __ocfs2_mknod_locked(dir, inode, dev, new_fe_bh,
-				    parent_fe_bh, handle, inode_ac,
+	mlog(0, "%s: claim lock block %llu for inode %llu\n",
+			osb->uuid_str,
+			(unsigned long long)fe_lock_blkno,
+			(unsigned long long)fe_blkno);
+
+	status = __ocfs2_mknod_locked(dir, inode, dev, new_fe_bh,
+				    parent_fe_bh, handle, inode_ac, lock_ac,
+				    fe_lock_blkno, suballoc_lock_loc, suballoc_lock_bit,
 				    fe_blkno, suballoc_loc, suballoc_bit);
+	if (status < 0) {
+		mlog_errno(status);
+		goto bail;
+	}
+
+	return 0;
+
+bail:
+	/* If ip_inode_lockres has been initialized, we need free
+	 * suballoc bits in iput() outside of this function.
+	 * Otherwise, it will cause a soft-readonly error. */
+	if (OCFS2_I(inode)->ip_inode_lockres.l_flags & OCFS2_LOCK_INITIALIZED)
+		return status;
+
+	if (lock_claimed) {
+		if (!suballoc_lock_loc)
+			suballoc_lock_loc = ocfs2_which_suballoc_group(fe_lock_blkno,
+					suballoc_lock_bit);
+
+		ocfs2_free_suballoc_bits(handle, lock_ac->ac_inode, lock_ac->ac_bh,
+				suballoc_lock_bit, suballoc_lock_loc, 1);
+	}
+
+	if (inode_claimed) {
+		if (!suballoc_loc)
+			suballoc_loc = ocfs2_which_suballoc_group(fe_blkno, suballoc_bit);
+
+		ocfs2_free_suballoc_bits(handle, inode_ac->ac_inode, inode_ac->ac_bh,
+				suballoc_bit, suballoc_loc, 1);
+	}
+
+	return status;
 }
 
 static int ocfs2_mkdir(struct inode *dir,
@@ -689,10 +883,25 @@ static int ocfs2_link(struct dentry *old_dentry,
 	sigset_t oldset;
 	u64 old_de_ino;
 
+#if defined(CONFIG_ARM64) || defined(CONFIG_ARM)
+	mlog(ML_ERROR, "%s not supported at ARM.", __func__);
+	return -ENOSYS;
+#endif
+
 	trace_ocfs2_link((unsigned long long)OCFS2_I(inode)->ip_blkno,
 			 old_dentry->d_name.len, old_dentry->d_name.name,
 			 dentry->d_name.len, dentry->d_name.name);
 
+	mlog(0, "%s: inode blkno %llu, old dentry %pd, dentry %pd\n",
+			osb->uuid_str, (unsigned long long)OCFS2_I(inode)->ip_blkno,
+			old_dentry, dentry);
+
+	if (ocfs2_test_invalid_fs(osb)) {
+		printk_once(KERN_ERR "%s: %s: OCFS2 invalid.\n",
+				 __func__, osb->uuid_str);
+		return -EIO;
+	}
+
 	if (S_ISDIR(inode->i_mode))
 		return -EPERM;
 
@@ -728,6 +937,7 @@ static int ocfs2_link(struct dentry *old_dentry,
 		goto out;
 	}
 
+	/* Check whether the file is exist, we might rm it on other nodes */
 	err = ocfs2_lookup_ino_from_name(old_dir, old_dentry->d_name.name,
 			old_dentry->d_name.len, &old_de_ino);
 	if (err) {
@@ -843,6 +1053,9 @@ static int ocfs2_remote_dentry_delete(struct dentry *dentry)
 {
 	int ret;
 
+	if (ocfs2_disk_lock(OCFS2_SB(dentry->d_sb)))
+		return 0;
+
 	ret = ocfs2_dentry_lock(dentry, 1);
 	if (ret)
 		mlog_errno(ret);
@@ -888,6 +1101,17 @@ static int ocfs2_unlink(struct inode *dir,
 			   (unsigned long long)OCFS2_I(dir)->ip_blkno,
 			   (unsigned long long)OCFS2_I(inode)->ip_blkno);
 
+	if (ocfs2_test_invalid_fs(osb)) {
+		printk_once(KERN_ERR "%s: %s: OCFS2 invalid.\n",
+				 __func__, osb->uuid_str);
+		return -EIO;
+	}
+
+	mlog(0, "%s: unlink filename %pd, inode %llu, nlink %u, parent dir inode %llu, nlink %u\n",
+			osb->uuid_str, dentry,
+			(unsigned long long)OCFS2_I(inode)->ip_blkno, inode->i_nlink,
+			(unsigned long long)OCFS2_I(dir)->ip_blkno, dir->i_nlink);
+
 	status = dquot_initialize(dir);
 	if (status) {
 		mlog_errno(status);
@@ -896,8 +1120,10 @@ static int ocfs2_unlink(struct inode *dir,
 
 	BUG_ON(d_inode(dentry->d_parent) != dir);
 
-	if (inode == osb->root_inode)
+	if (inode == osb->root_inode) {
+		mlog_errno(-EPERM);
 		return -EPERM;
+	}
 
 	status = ocfs2_inode_lock_nested(dir, &parent_node_bh, 1,
 					 OI_LS_PARENT);
@@ -941,6 +1167,7 @@ static int ocfs2_unlink(struct inode *dir,
 		}
 	}
 
+	/* to prevent other nodes still have inode cache*/
 	status = ocfs2_remote_dentry_delete(dentry);
 	if (status < 0) {
 		/* This remote delete should succeed under all normal
@@ -984,6 +1211,11 @@ static int ocfs2_unlink(struct inode *dir,
 		mlog_errno(status);
 		goto leave;
 	}
+	mlog(ML_DEBUG,
+			"%s: delete_entry %pd, dir %llu, nlink %u, inode %llu, nlink %u\n",
+			osb->uuid_str, dentry,
+			(unsigned long long)OCFS2_I(dir)->ip_blkno, dir->i_nlink,
+			(unsigned long long)OCFS2_I(inode)->ip_blkno, inode->i_nlink);
 
 	if (S_ISDIR(inode->i_mode))
 		drop_nlink(inode);
@@ -1008,6 +1240,10 @@ static int ocfs2_unlink(struct inode *dir,
 				orphan_name, &orphan_insert, orphan_dir, false);
 		if (status < 0)
 			mlog_errno(status);
+		else
+			mlog(ML_DEBUG, "%s: file %pd move to orphan %s, inode %llu\n",
+					osb->uuid_str, dentry,
+					orphan_name, (unsigned long long)OCFS2_I(inode)->ip_blkno);
 	}
 
 leave:
@@ -1035,9 +1271,14 @@ static int ocfs2_unlink(struct inode *dir,
 	if (status && (status != -ENOTEMPTY) && (status != -ENOENT))
 		mlog_errno(status);
 
+	mlog(ML_DEBUG, "%s: %s completed, status %d\n", osb->uuid_str, __func__, status);
 	return status;
 }
 
+/* return 1 if dest inode is the ancestor of src inode,
+ * 0 if not,
+ * or <0 if error occurs
+ */
 static int ocfs2_check_if_ancestor(struct ocfs2_super *osb,
 		u64 src_inode_no, u64 dest_inode_no)
 {
@@ -1076,6 +1317,7 @@ static int ocfs2_check_if_ancestor(struct ocfs2_super *osb,
 			break;
 		}
 
+		/* reach to root directory */
 		if (parent_inode_no == osb->root_inode->i_ino) {
 			ret = 0;
 			break;
@@ -1083,10 +1325,11 @@ static int ocfs2_check_if_ancestor(struct ocfs2_super *osb,
 
 		child_inode_no = parent_inode_no;
 
+		/* avoid the case of endless loop */
 		if (++i >= MAX_LOOKUP_TIMES) {
-			mlog(ML_NOTICE, "max lookup times reached, filesystem "
-					"may have nested directories, "
-					"src inode: %llu, dest inode: %llu.\n",
+			mlog(ML_NOTICE,
+					"%s: max lookup times (32) reached, filesystem may have nested directories, src inode: %llu, dest inode: %llu.\n",
+					osb->uuid_str,
 					(unsigned long long)src_inode_no,
 					(unsigned long long)dest_inode_no);
 			ret = 0;
@@ -1121,8 +1364,8 @@ static int ocfs2_double_lock(struct ocfs2_super *osb,
 	if (*bh2)
 		*bh2 = NULL;
 
-	/* we always want to lock the one with the lower lockid first.
-	 * and if they are nested, we lock ancestor first */
+	/* we always want to lock the one with the lower lockid first. */
+	/* if the lower inode no is child, we lock ancestor first */
 	if (oi1->ip_blkno != oi2->ip_blkno) {
 		inode1_is_ancestor = ocfs2_check_if_ancestor(osb, oi2->ip_blkno,
 				oi1->ip_blkno);
@@ -1157,7 +1400,7 @@ static int ocfs2_double_lock(struct ocfs2_super *osb,
 
 	/* lock id1 */
 	status = ocfs2_inode_lock_nested(inode1, bh1, 1,
-			rename == 1 ?  OI_LS_RENAME2 : OI_LS_PARENT);
+			rename == 1 ? OI_LS_RENAME2 : OI_LS_PARENT);
 	if (status < 0) {
 		/*
 		 * An error return must mean that no cluster locks
@@ -1211,7 +1454,7 @@ static int ocfs2_rename(struct inode *old_dir,
 	handle_t *handle = NULL;
 	struct buffer_head *old_dir_bh = NULL;
 	struct buffer_head *new_dir_bh = NULL;
-	u32 old_dir_nlink = old_dir->i_nlink;
+	nlink_t old_dir_nlink = old_dir->i_nlink;
 	struct ocfs2_dinode *old_di;
 	struct ocfs2_dir_lookup_result old_inode_dot_dot_res = { NULL, };
 	struct ocfs2_dir_lookup_result target_lookup_res = { NULL, };
@@ -1243,6 +1486,21 @@ static int ocfs2_rename(struct inode *old_dir,
 
 	osb = OCFS2_SB(old_dir->i_sb);
 
+	if (ocfs2_test_invalid_fs(osb)) {
+		printk_once(KERN_ERR "%s: %s: OCFS2 invalid.\n",
+				__func__, osb->uuid_str);
+		return -EIO;
+	}
+
+	mlog(0, "%s: %pd -> %pd, old dir %llu, nlink %u, old inode %llu, "
+			"nlink %u, new dir %llu, nlink %u, new inode %llu, nlink %u.\n",
+			osb->uuid_str, old_dentry, new_dentry,
+			(unsigned long long)OCFS2_I(old_dir)->ip_blkno, old_dir->i_nlink,
+			(unsigned long long)OCFS2_I(old_inode)->ip_blkno, old_inode->i_nlink,
+			(unsigned long long)OCFS2_I(new_dir)->ip_blkno, new_dir->i_nlink,
+			(unsigned long long)((!new_inode) ? 0 : OCFS2_I(new_inode)->ip_blkno),
+			(!new_inode) ? 0 : new_inode->i_nlink);
+
 	if (new_inode) {
 		if (!igrab(new_inode))
 			BUG();
@@ -1267,6 +1525,16 @@ static int ocfs2_rename(struct inode *old_dir,
 		}
 		rename_lock = 1;
 
+		if (new_dir->i_ino == old_inode->i_ino ||
+				old_dir->i_ino == old_inode->i_ino) {
+			status = -EACCES;
+			mlog(ML_ERROR,
+					"%s: old inode %lu should not be equal to new dir inode %lu or old dir inode %lu\n",
+					osb->uuid_str, old_inode->i_ino,
+					new_dir->i_ino, old_dir->i_ino);
+			goto bail;
+		}
+
 		/* here we cannot guarantee the inodes haven't just been
 		 * changed, so check if they are nested again */
 		status = ocfs2_check_if_ancestor(osb, new_dir->i_ino,
@@ -1275,10 +1543,11 @@ static int ocfs2_rename(struct inode *old_dir,
 			mlog_errno(status);
 			goto bail;
 		} else if (status == 1) {
-			status = -EPERM;
-			trace_ocfs2_rename_not_permitted(
-					(unsigned long long)old_inode->i_ino,
-					(unsigned long long)new_dir->i_ino);
+			status = -EACCES;
+			mlog(ML_ERROR,
+					"%s: old inode %lu should not be ancestor of new dir inode %lu\n",
+					osb->uuid_str,
+					old_inode->i_ino, new_dir->i_ino);
 			goto bail;
 		}
 	}
@@ -1293,6 +1562,10 @@ static int ocfs2_rename(struct inode *old_dir,
 	parents_locked = 1;
 
 	if (!new_dir->i_nlink) {
+		mlog(ML_ERROR,
+				"%s: new dir %lu has been removed, inode %lu can not be moved into it.\n",
+				osb->uuid_str,
+				new_dir->i_ino, old_inode->i_ino);
 		status = -EACCES;
 		goto bail;
 	}
@@ -1304,7 +1577,7 @@ static int ocfs2_rename(struct inode *old_dir,
 			new_dir_bh = old_dir_bh;
 			get_bh(new_dir_bh);
 		} else {
-			mlog(ML_ERROR, "no old_dir_bh!\n");
+			mlog(ML_ERROR, "%s: no old_dir_bh!\n", osb->uuid_str);
 			status = -EIO;
 			goto bail;
 		}
@@ -1522,14 +1795,20 @@ static int ocfs2_rename(struct inode *old_dir,
 					&orphan_insert, orphan_dir, false);
 			if (status < 0) {
 				mlog_errno(status);
+				if (!ocfs2_test_invalid_fs(osb))
+					o2hb_handle_invalid(osb->uuid_str);
 				goto bail;
 			}
+			mlog(ML_DEBUG, "%s: new inode %llu add to orphan dir\n",
+				osb->uuid_str, (unsigned long long)OCFS2_I(new_inode)->ip_blkno);
 		}
 	} else {
 		/* if the name was not found in new_dir, add it now */
 		status = ocfs2_add_entry(handle, new_dentry, old_inode,
 					 OCFS2_I(old_inode)->ip_blkno,
 					 new_dir_bh, &target_insert);
+		if (status < 0)
+			mlog_errno(status);
 	}
 
 	old_inode->i_ctime = current_time(old_inode);
@@ -1558,24 +1837,23 @@ static int ocfs2_rename(struct inode *old_dir,
 				  old_dentry->d_name.len, old_dir,
 				  &old_entry_lookup);
 	if (status) {
-		if (!is_journal_aborted(osb->journal->j_journal)) {
-			ocfs2_error(osb->sb, "new entry %.*s is added, but old entry %.*s "
-					"is not deleted.",
-					new_dentry->d_name.len, new_dentry->d_name.name,
-					old_dentry->d_name.len, old_dentry->d_name.name);
-		}
+		mlog(ML_ERROR,
+				"%s: new entry %pd is added, but old entry %pd is not deleted.\n",
+				osb->uuid_str, new_dentry, old_dentry);
+		if (!ocfs2_test_invalid_fs(osb))
+			o2hb_handle_invalid(osb->uuid_str);
+
 		goto bail;
 	}
 
 	status = ocfs2_delete_entry(handle, old_dir, &old_entry_lookup);
 	if (status < 0) {
-		mlog_errno(status);
-		if (!is_journal_aborted(osb->journal->j_journal)) {
-			ocfs2_error(osb->sb, "new entry %.*s is added, but old entry %.*s "
-					"is not deleted.",
-					new_dentry->d_name.len, new_dentry->d_name.name,
-					old_dentry->d_name.len, old_dentry->d_name.name);
-		}
+		mlog(ML_ERROR,
+				"%s: new entry %pd is added, but old entry %pd is not deleted.\n",
+				osb->uuid_str, new_dentry, old_dentry);
+		if (!ocfs2_test_invalid_fs(osb))
+			o2hb_handle_invalid(osb->uuid_str);
+
 		goto bail;
 	}
 
@@ -1588,6 +1866,8 @@ static int ocfs2_rename(struct inode *old_dir,
 	if (update_dot_dot) {
 		status = ocfs2_update_entry(old_inode, handle,
 					    &old_inode_dot_dot_res, new_dir);
+		if (status < 0)
+			mlog_errno(status);
 		drop_nlink(old_dir);
 		if (new_inode) {
 			drop_nlink(new_inode);
@@ -1616,8 +1896,9 @@ static int ocfs2_rename(struct inode *old_dir,
 
 	if (old_dir_nlink != old_dir->i_nlink) {
 		if (!old_dir_bh) {
-			mlog(ML_ERROR, "need to change nlink for old dir "
-			     "%llu from %d to %d but bh is NULL!\n",
+			mlog(ML_ERROR,
+			     "%s: need to change nlink for old dir %llu from %d to %d but bh is NULL!\n",
+			     osb->uuid_str,
 			     (unsigned long long)OCFS2_I(old_dir)->ip_blkno,
 			     (int)old_dir_nlink, old_dir->i_nlink);
 		} else {
@@ -1626,6 +1907,8 @@ static int ocfs2_rename(struct inode *old_dir,
 							 INODE_CACHE(old_dir),
 							 old_dir_bh,
 							 OCFS2_JOURNAL_ACCESS_WRITE);
+			if (status < 0)
+				mlog_errno(status);
 			fe = (struct ocfs2_dinode *) old_dir_bh->b_data;
 			ocfs2_set_links_count(fe, old_dir->i_nlink);
 			ocfs2_journal_dirty(handle, old_dir_bh);
@@ -1672,7 +1955,7 @@ static int ocfs2_rename(struct inode *old_dir,
 	brelse(old_dir_bh);
 	brelse(new_dir_bh);
 
-	if (status)
+	if (status && (status != -ENOTEMPTY))
 		mlog_errno(status);
 
 	return status;
@@ -1797,6 +2080,7 @@ static int ocfs2_symlink(struct inode *dir,
 	struct ocfs2_alloc_context *inode_ac = NULL;
 	struct ocfs2_alloc_context *data_ac = NULL;
 	struct ocfs2_alloc_context *xattr_ac = NULL;
+	struct ocfs2_alloc_context *lock_ac = NULL;
 	int want_clusters = 0;
 	int xattr_credits = 0;
 	struct ocfs2_security_xattr_info si = {
@@ -1807,6 +2091,12 @@ static int ocfs2_symlink(struct inode *dir,
 	sigset_t oldset;
 	int did_block_signals = 0;
 	struct ocfs2_dentry_lock *dl = NULL;
+	bool disk_lock;
+
+#if defined(CONFIG_ARM64) || defined(CONFIG_ARM)
+	mlog(ML_ERROR, "%s not supported at ARM.", __func__);
+	return -ENOSYS;
+#endif
 
 	trace_ocfs2_symlink_begin(dir, dentry, symname,
 				  dentry->d_name.len, dentry->d_name.name);
@@ -1820,9 +2110,16 @@ static int ocfs2_symlink(struct inode *dir,
 	sb = dir->i_sb;
 	osb = OCFS2_SB(sb);
 
+	if (ocfs2_test_invalid_fs(osb)) {
+		printk_once(KERN_ERR "%s: %s: OCFS2 invalid.\n",
+				__func__, osb->uuid_str);
+		return -EIO;
+	}
+
 	l = strlen(symname) + 1;
 
-	credits = ocfs2_calc_symlink_credits(sb);
+	disk_lock = ocfs2_disk_lock_support(osb);
+	credits = ocfs2_calc_symlink_credits(sb, disk_lock);
 
 	/* lock the parent directory */
 	status = ocfs2_inode_lock(dir, &parent_fe_bh, 1);
@@ -1852,6 +2149,15 @@ static int ocfs2_symlink(struct inode *dir,
 		goto bail;
 	}
 
+	if (disk_lock) {
+		status = ocfs2_reserve_new_lock(osb, &lock_ac);
+		if (status < 0) {
+			if (status != -ENOSPC)
+				mlog_errno(status);
+			goto bail;
+		}
+	}
+
 	status = ocfs2_reserve_new_inode(osb, &inode_ac);
 	if (status < 0) {
 		if (status != -ENOSPC)
@@ -1923,7 +2229,7 @@ static int ocfs2_symlink(struct inode *dir,
 
 	status = ocfs2_mknod_locked(osb, dir, inode,
 				    0, &new_fe_bh, parent_fe_bh, handle,
-				    inode_ac);
+				    inode_ac, lock_ac, disk_lock);
 	if (status < 0) {
 		mlog_errno(status);
 		goto bail;
@@ -1950,8 +2256,8 @@ static int ocfs2_symlink(struct inode *dir,
 		if (status < 0) {
 			if (status != -ENOSPC && status != -EINTR) {
 				mlog(ML_ERROR,
-				     "Failed to extend file to %llu\n",
-				     (unsigned long long)newsize);
+				     "%s: Failed to extend file to %llu\n",
+				     osb->uuid_str, (unsigned long long)newsize);
 				mlog_errno(status);
 				status = -ENOSPC;
 			}
@@ -2039,6 +2345,8 @@ static int ocfs2_symlink(struct inode *dir,
 		ocfs2_free_alloc_context(data_ac);
 	if (xattr_ac)
 		ocfs2_free_alloc_context(xattr_ac);
+	if (lock_ac)
+		ocfs2_free_alloc_context(lock_ac);
 	if ((status < 0) && inode) {
 		if (dl)
 			ocfs2_cleanup_add_entry_failure(osb, dentry, inode);
@@ -2085,11 +2393,13 @@ static int ocfs2_blkno_stringify(u64 blkno, char *name)
 
 static int ocfs2_lookup_lock_orphan_dir(struct ocfs2_super *osb,
 					struct inode **ret_orphan_dir,
-					struct buffer_head **ret_orphan_dir_bh)
+					struct buffer_head **ret_orphan_dir_bh,
+					bool dio)
 {
 	struct inode *orphan_dir_inode;
 	struct buffer_head *orphan_dir_bh = NULL;
 	int ret = 0;
+	int flags = dio ? OCFS2_LOCK_NONBLOCK : 0;
 
 	orphan_dir_inode = ocfs2_get_system_file_inode(osb,
 						       ORPHAN_DIR_SYSTEM_INODE,
@@ -2101,13 +2411,21 @@ static int ocfs2_lookup_lock_orphan_dir(struct ocfs2_super *osb,
 	}
 
 	inode_lock(orphan_dir_inode);
-
-	ret = ocfs2_inode_lock(orphan_dir_inode, &orphan_dir_bh, 1);
+	/* This is helping work around a lock inversion between inode
+	 * and orphan dir inode lock. The orphan scan holds the orphan dir
+	 * inode lock, and wait for inode lock if inode state is I_NEW. So
+	 * use OCFS2_LOCK_NONBLOCK and in case of -EAGAIN, unlock inode
+	 * first and retry */
+	ret = ocfs2_inode_lock_full(orphan_dir_inode, &orphan_dir_bh, 1, flags);
 	if (ret < 0) {
 		inode_unlock(orphan_dir_inode);
-		iput(orphan_dir_inode);
+		if (ret == -EAGAIN && dio)
+			*ret_orphan_dir = orphan_dir_inode;
+		else {
+			iput(orphan_dir_inode);
+			mlog_errno(ret);
+		}
 
-		mlog_errno(ret);
 		return ret;
 	}
 
@@ -2185,9 +2503,12 @@ static int ocfs2_prepare_orphan_dir(struct ocfs2_super *osb,
 	int ret = 0;
 
 	ret = ocfs2_lookup_lock_orphan_dir(osb, &orphan_dir_inode,
-					   &orphan_dir_bh);
+					   &orphan_dir_bh, dio);
 	if (ret < 0) {
-		mlog_errno(ret);
+		if (ret == -EAGAIN && dio)
+			*ret_orphan_dir = orphan_dir_inode;
+		else
+			mlog_errno(ret);
 		return ret;
 	}
 
@@ -2254,6 +2575,8 @@ static int ocfs2_orphan_add(struct ocfs2_super *osb,
 	 * It's safe anyway, though some callers may duplicate the journaling.
 	 * Journaling within the func just make the logic look more
 	 * straightforward.
+	 * Notify intent to modify a buffer firstly,
+	 * if notify failed, file(inode) can not be inserted into orphan_dir.
 	 */
 	status = ocfs2_journal_access_di(handle,
 					 INODE_CACHE(inode),
@@ -2302,6 +2625,10 @@ static int ocfs2_orphan_add(struct ocfs2_super *osb,
 	trace_ocfs2_orphan_add_end((unsigned long long)OCFS2_I(inode)->ip_blkno,
 				   osb->slot_num);
 
+	/*
+	 * if insert file into orphan_dir failed, orphan_fe->i_links_count_hi
+	 * and orphan_dir_inode->i_nlink should be restored to their original.
+	 */
 rollback:
 	if (status < 0) {
 		if (S_ISDIR(inode->i_mode))
@@ -2315,6 +2642,48 @@ static int ocfs2_orphan_add(struct ocfs2_super *osb,
 	return status;
 }
 
+int ocfs2_remove_invalid_entry(struct ocfs2_super *osb, u64 ino,
+		struct inode *orphan_dir_inode)
+{
+	handle_t *handle = NULL;
+	char name[OCFS2_ORPHAN_NAMELEN + 1];
+	struct ocfs2_dir_lookup_result lookup = { NULL, };
+	int status = 0;
+
+	status = ocfs2_blkno_stringify(ino, name);
+	if (status < 0) {
+		mlog_errno(status);
+		return status;
+	}
+
+	handle = ocfs2_start_trans(osb,
+			OCFS2_INODE_DEL_FROM_ORPHAN_CREDITS);
+	if (IS_ERR(handle)) {
+		status = PTR_ERR(handle);
+		return status;
+	}
+
+	/* find it's spot in the orphan directory */
+	status = ocfs2_find_entry(name, strlen(name), orphan_dir_inode,
+			&lookup);
+	if (status) {
+		mlog_errno(status);
+		goto leave;
+	}
+
+	/* remove it from the orphan directory */
+	status = ocfs2_delete_entry(handle, orphan_dir_inode, &lookup);
+	if (status < 0)
+		mlog_errno(status);
+
+leave:
+	ocfs2_free_dir_lookup_result(&lookup);
+
+	ocfs2_commit_trans(osb, handle);
+
+	return status;
+}
+
 /* unlike orphan_add, we expect the orphan dir to already be locked here. */
 int ocfs2_orphan_del(struct ocfs2_super *osb,
 		     handle_t *handle,
@@ -2384,8 +2753,6 @@ int ocfs2_orphan_del(struct ocfs2_super *osb,
 leave:
 	ocfs2_free_dir_lookup_result(&lookup);
 
-	if (status)
-		mlog_errno(status);
 	return status;
 }
 
@@ -2429,7 +2796,7 @@ static int ocfs2_prep_new_orphaned_file(struct inode *dir,
 	struct buffer_head *orphan_dir_bh = NULL;
 	struct ocfs2_alloc_context *inode_ac = NULL;
 
-	ret = ocfs2_lookup_lock_orphan_dir(osb, &orphan_dir, &orphan_dir_bh);
+	ret = ocfs2_lookup_lock_orphan_dir(osb, &orphan_dir, &orphan_dir_bh, false);
 	if (ret < 0) {
 		mlog_errno(ret);
 		return ret;
@@ -2497,8 +2864,9 @@ int ocfs2_create_inode_in_orphan(struct inode *dir,
 	struct buffer_head *new_di_bh = NULL;
 	struct ocfs2_alloc_context *inode_ac = NULL;
 	struct ocfs2_dir_lookup_result orphan_insert = { NULL, };
-	u64 di_blkno, suballoc_loc;
+	u64 di_blkno = 0UL, suballoc_loc;
 	u16 suballoc_bit;
+	bool disk_lock = ocfs2_disk_lock_support(osb);
 
 	status = ocfs2_inode_lock(dir, &parent_di_bh, 1);
 	if (status < 0) {
@@ -2524,7 +2892,8 @@ int ocfs2_create_inode_in_orphan(struct inode *dir,
 		goto leave;
 	}
 
-	handle = ocfs2_start_trans(osb, ocfs2_mknod_credits(osb->sb, 0, 0));
+	handle = ocfs2_start_trans(osb,
+			ocfs2_mknod_credits(osb->sb, 0, 0, disk_lock));
 	if (IS_ERR(handle)) {
 		status = PTR_ERR(handle);
 		handle = NULL;
@@ -2549,7 +2918,7 @@ int ocfs2_create_inode_in_orphan(struct inode *dir,
 	/* do the real work now. */
 	status = __ocfs2_mknod_locked(dir, inode,
 				      0, &new_di_bh, parent_di_bh, handle,
-				      inode_ac, di_blkno, suballoc_loc,
+				      inode_ac, NULL, 0, 0, 0, di_blkno, suballoc_loc,
 				      suballoc_bit);
 	if (status < 0) {
 		mlog_errno(status);
diff --git a/fs/ocfs2/namei.h b/fs/ocfs2/namei.h
index cc091ed02b4a..5fafa7cd4a36 100644
--- a/fs/ocfs2/namei.h
+++ b/fs/ocfs2/namei.h
@@ -25,6 +25,8 @@ int ocfs2_orphan_del(struct ocfs2_super *osb,
 		     struct inode *inode,
 		     struct buffer_head *orphan_dir_bh,
 		     bool dio);
+int ocfs2_remove_invalid_entry(struct ocfs2_super *osb, u64 ino,
+		struct inode *orphan_dir_inode);
 int ocfs2_create_inode_in_orphan(struct inode *dir,
 				 int mode,
 				 struct inode **new_inode);
diff --git a/fs/ocfs2/quota_global.c b/fs/ocfs2/quota_global.c
index eda83487c9ec..f96b503f9297 100644
--- a/fs/ocfs2/quota_global.c
+++ b/fs/ocfs2/quota_global.c
@@ -24,13 +24,21 @@
 #include "journal.h"
 #include "file.h"
 #include "sysfile.h"
-#include "dlmglue.h"
+#include "lockglue.h"
 #include "uptodate.h"
 #include "super.h"
 #include "buffer_head_io.h"
 #include "quota.h"
 #include "ocfs2_trace.h"
 
+#ifndef llist_for_each_entry_safe
+#define llist_for_each_entry_safe(pos, n, node, member)                        \
+		for (pos = llist_entry((node), typeof(*pos), member);                  \
+			 &pos->member != NULL &&                                           \
+			    (n = llist_entry(pos->member.next, typeof(*n), member), true); \
+			 pos = n)
+#endif
+
 /*
  * Locking of quotas with OCFS2 is rather complex. Here are rules that
  * should be obeyed by all the functions:
@@ -116,10 +124,8 @@ static int ocfs2_global_is_id(void *dp, struct dquot *dquot)
 	struct ocfs2_global_disk_dqblk *d = dp;
 	struct ocfs2_mem_dqinfo *oinfo =
 			sb_dqinfo(dquot->dq_sb, dquot->dq_id.type)->dqi_priv;
-
 	if (qtree_entry_unused(&oinfo->dqi_gi, dp))
 		return 0;
-
 	return qid_eq(make_kqid(&init_user_ns, dquot->dq_id.type,
 				le32_to_cpu(d->dqb_id)),
 		      dquot->dq_id);
@@ -737,7 +743,11 @@ static int ocfs2_release_dquot(struct dquot *dquot)
 		 * dropping.  Quota code rechecks after calling
 		 * ->release_dquot() and won't free dquot structure.
 		 */
-		dqgrab(dquot);
+		/* Make sure someone else has active reference to dquot */
+		WARN_ON_ONCE(!atomic_read(&dquot->dq_count));
+		WARN_ON_ONCE(!test_bit(DQ_ACTIVE_B, &dquot->dq_flags));
+		atomic_inc(&dquot->dq_count);
+
 		/* First entry on list -> queue work */
 		if (llist_add(&OCFS2_DQUOT(dquot)->list, &osb->dquot_drop_list))
 			queue_work(osb->ocfs2_wq, &osb->dquot_drop_work);
diff --git a/fs/ocfs2/quota_local.c b/fs/ocfs2/quota_local.c
index b1a8b046f4c2..ff393449dff3 100644
--- a/fs/ocfs2/quota_local.c
+++ b/fs/ocfs2/quota_local.c
@@ -19,7 +19,7 @@
 #include "buffer_head_io.h"
 #include "journal.h"
 #include "sysfile.h"
-#include "dlmglue.h"
+#include "lockglue.h"
 #include "quota.h"
 #include "uptodate.h"
 #include "super.h"
@@ -137,13 +137,14 @@ static int ocfs2_read_quota_block(struct inode *inode, u64 v_block,
 	int rc = 0;
 	struct buffer_head *tmp = *bh;
 
-	if (i_size_read(inode) >> inode->i_sb->s_blocksize_bits <= v_block)
-		return ocfs2_error(inode->i_sb,
-				"Quota file %llu is probably corrupted! Requested to read block %Lu but file has size only %Lu\n",
-				(unsigned long long)OCFS2_I(inode)->ip_blkno,
-				(unsigned long long)v_block,
-				(unsigned long long)i_size_read(inode));
-
+	if (i_size_read(inode) >> inode->i_sb->s_blocksize_bits <= v_block) {
+		ocfs2_error(inode->i_sb,
+			    "Quota file %llu is probably corrupted! Requested to read block %llu but file has size only %llu\n",
+			    (unsigned long long)OCFS2_I(inode)->ip_blkno,
+			    (unsigned long long)v_block,
+			    (unsigned long long)i_size_read(inode));
+		return -EROFS;
+	}
 	rc = ocfs2_read_virt_blocks(inode, v_block, 1, &tmp, 0,
 				    ocfs2_validate_quota_block);
 	if (rc)
@@ -398,9 +399,8 @@ struct ocfs2_quota_recovery *ocfs2_begin_quota_recovery(
 	int status = 0;
 	struct ocfs2_quota_recovery *rec;
 
-	printk(KERN_NOTICE "ocfs2: Beginning quota recovery on device (%s) for "
-	       "slot %u\n", osb->dev_str, slot_num);
-
+	mlog(ML_NOTICE, "%s: Beginning quota recovery in slot %u\n",
+			osb->uuid_str, slot_num);
 	rec = ocfs2_alloc_quota_recovery();
 	if (!rec)
 		return ERR_PTR(-ENOMEM);
@@ -594,9 +594,8 @@ int ocfs2_finish_quota_recovery(struct ocfs2_super *osb,
 	struct inode *lqinode;
 	unsigned int flags;
 
-	printk(KERN_NOTICE "ocfs2: Finishing quota recovery on device (%s) for "
-	       "slot %u\n", osb->dev_str, slot_num);
-
+	mlog(ML_NOTICE, "%s: Finishing quota recovery in slot %u\n",
+			osb->uuid_str, slot_num);
 	down_read(&sb->s_umount);
 	for (type = 0; type < OCFS2_MAXQUOTAS; type++) {
 		if (list_empty(&(rec->r_list[type])))
@@ -612,9 +611,9 @@ int ocfs2_finish_quota_recovery(struct ocfs2_super *osb,
 		/* Someone else is holding the lock? Then he must be
 		 * doing the recovery. Just skip the file... */
 		if (status == -EAGAIN) {
-			printk(KERN_NOTICE "ocfs2: Skipping quota recovery on "
-			       "device (%s) for slot %d because quota file is "
-			       "locked.\n", osb->dev_str, slot_num);
+			mlog(ML_NOTICE,
+			     "%s: skipping quota recovery for slot %d because quota file is locked.\n",
+			     osb->uuid_str, slot_num);
 			status = 0;
 			goto out_put;
 		} else if (status < 0) {
@@ -695,8 +694,7 @@ static int ocfs2_local_read_info(struct super_block *sb, int type)
 	info->dqi_max_ino_limit = 0x7fffffffffffffffLL;
 	oinfo = kmalloc(sizeof(struct ocfs2_mem_dqinfo), GFP_NOFS);
 	if (!oinfo) {
-		mlog(ML_ERROR, "failed to allocate memory for ocfs2 quota"
-			       " info.");
+		mlog(ML_ERROR, "failed to allocate memory for ocfs2 quota info.\n");
 		goto out_err;
 	}
 	info->dqi_priv = oinfo;
@@ -774,7 +772,7 @@ static int ocfs2_local_read_info(struct super_block *sb, int type)
 	if (oinfo) {
 		iput(oinfo->dqi_gqinode);
 		ocfs2_simple_drop_lockres(OCFS2_SB(sb), &oinfo->dqi_gqlock);
-		ocfs2_lock_res_free(&oinfo->dqi_gqlock);
+		ocfs2_lock_res_free(OCFS2_SB(sb), &oinfo->dqi_gqlock);
 		brelse(oinfo->dqi_lqi_bh);
 		if (locked)
 			ocfs2_inode_unlock(lqinode, 1);
@@ -815,7 +813,7 @@ static int ocfs2_local_free_info(struct super_block *sb, int type)
 
 	iput(oinfo->dqi_gqinode);
 	ocfs2_simple_drop_lockres(OCFS2_SB(sb), &oinfo->dqi_gqlock);
-	ocfs2_lock_res_free(&oinfo->dqi_gqlock);
+	ocfs2_lock_res_free(OCFS2_SB(sb), &oinfo->dqi_gqlock);
 	list_for_each_entry(chunk, &oinfo->dqi_chunk, qc_chunk) {
 		dchunk = (struct ocfs2_local_disk_chunk *)
 					(chunk->qc_headerbh->b_data);
diff --git a/fs/ocfs2/refcounttree.c b/fs/ocfs2/refcounttree.c
index 3b397fa9c9e8..a9d2c1764b1e 100644
--- a/fs/ocfs2/refcounttree.c
+++ b/fs/ocfs2/refcounttree.c
@@ -20,7 +20,7 @@
 #include "blockcheck.h"
 #include "refcounttree.h"
 #include "sysfile.h"
-#include "dlmglue.h"
+#include "lockglue.h"
 #include "extent_map.h"
 #include "aops.h"
 #include "xattr.h"
@@ -95,30 +95,30 @@ static int ocfs2_validate_refcount_block(struct super_block *sb,
 
 
 	if (!OCFS2_IS_VALID_REFCOUNT_BLOCK(rb)) {
-		rc = ocfs2_error(sb,
-				 "Refcount block #%llu has bad signature %.*s\n",
+		ocfs2_error(sb,
+				 "Refcount block #%llu has bad signature %.*s",
 				 (unsigned long long)bh->b_blocknr, 7,
 				 rb->rf_signature);
-		goto out;
+		return -EINVAL;
 	}
 
 	if (le64_to_cpu(rb->rf_blkno) != bh->b_blocknr) {
-		rc = ocfs2_error(sb,
-				 "Refcount block #%llu has an invalid rf_blkno of %llu\n",
+		ocfs2_error(sb,
+				 "Refcount block #%llu has an invalid rf_blkno of %llu",
 				 (unsigned long long)bh->b_blocknr,
 				 (unsigned long long)le64_to_cpu(rb->rf_blkno));
-		goto out;
+		return -EINVAL;
 	}
 
 	if (le32_to_cpu(rb->rf_fs_generation) != OCFS2_SB(sb)->fs_generation) {
-		rc = ocfs2_error(sb,
-				 "Refcount block #%llu has an invalid rf_fs_generation of #%u\n",
+		ocfs2_error(sb,
+				 "Refcount block #%llu has an invalid rf_fs_generation of #%u",
 				 (unsigned long long)bh->b_blocknr,
 				 le32_to_cpu(rb->rf_fs_generation));
-		goto out;
+		return -EINVAL;
 	}
-out:
-	return rc;
+
+	return 0;
 }
 
 static int ocfs2_read_refcount_block(struct ocfs2_caching_info *ci,
@@ -247,7 +247,7 @@ static void ocfs2_free_refcount_tree(struct ocfs2_refcount_tree *tree)
 {
 	ocfs2_metadata_cache_exit(&tree->rf_ci);
 	ocfs2_simple_drop_lockres(OCFS2_SB(tree->rf_sb), &tree->rf_lockres);
-	ocfs2_lock_res_free(&tree->rf_lockres);
+	ocfs2_lock_res_free(OCFS2_SB(tree->rf_sb), &tree->rf_lockres);
 	kfree(tree);
 }
 
@@ -610,6 +610,7 @@ static int ocfs2_create_refcount_tree(struct inode *inode,
 		mlog_errno(ret);
 		goto out_commit;
 	}
+
 	ocfs2_set_new_buffer_uptodate(&new_tree->rf_ci, new_bh);
 
 	ret = ocfs2_journal_access_rb(handle, &new_tree->rf_ci, new_bh,
@@ -1063,7 +1064,7 @@ static int ocfs2_get_refcount_rec(struct ocfs2_caching_info *ci,
 				  struct buffer_head **ret_bh)
 {
 	int ret = 0, i, found;
-	u32 low_cpos, cpos_end;
+	u32 low_cpos, cpos_end = 0;
 	struct ocfs2_extent_list *el;
 	struct ocfs2_extent_rec *rec = NULL;
 	struct ocfs2_extent_block *eb = NULL;
@@ -1094,10 +1095,11 @@ static int ocfs2_get_refcount_rec(struct ocfs2_caching_info *ci,
 		el = &eb->h_list;
 
 		if (el->l_tree_depth) {
-			ret = ocfs2_error(sb,
+			ocfs2_error(sb,
 					  "refcount tree %llu has non zero tree depth in leaf btree tree block %llu\n",
 					  (unsigned long long)ocfs2_metadata_cache_owner(ci),
 					  (unsigned long long)eb_bh->b_blocknr);
+			ret = -EROFS;
 			goto out;
 		}
 	}
@@ -2348,8 +2350,10 @@ static int ocfs2_mark_extent_refcounted(struct inode *inode,
 					   cpos, len, phys);
 
 	if (!ocfs2_refcount_tree(OCFS2_SB(inode->i_sb))) {
-		ret = ocfs2_error(inode->i_sb, "Inode %lu want to use refcount tree, but the feature bit is not set in the super block\n",
+		ocfs2_error(inode->i_sb,
+				  "Inode %lu want to use refcount tree, but the feature bit is not set in the super block.",
 				  inode->i_ino);
+		ret = -EROFS;
 		goto out;
 	}
 
@@ -2531,8 +2535,10 @@ int ocfs2_prepare_refcount_change_for_del(struct inode *inode,
 	u64 start_cpos = ocfs2_blocks_to_clusters(inode->i_sb, phys_blkno);
 
 	if (!ocfs2_refcount_tree(OCFS2_SB(inode->i_sb))) {
-		ret = ocfs2_error(inode->i_sb, "Inode %lu want to use refcount tree, but the feature bit is not set in the super block\n",
+		ocfs2_error(inode->i_sb,
+				  "Inode %lu want to use refcount tree, but the feature bit is not set in the super block.",
 				  inode->i_ino);
+		ret = -EROFS;
 		goto out;
 	}
 
@@ -2656,10 +2662,11 @@ static int ocfs2_refcount_cal_cow_clusters(struct inode *inode,
 		el = &eb->h_list;
 
 		if (el->l_tree_depth) {
-			ret = ocfs2_error(inode->i_sb,
+			ocfs2_error(inode->i_sb,
 					  "Inode %lu has non zero tree depth in leaf block %llu\n",
 					  inode->i_ino,
 					  (unsigned long long)eb_bh->b_blocknr);
+			ret = -EROFS;
 			goto out;
 		}
 	}
@@ -3094,9 +3101,10 @@ static int ocfs2_clear_ext_refcount(handle_t *handle,
 
 	index = ocfs2_search_extent_list(el, cpos);
 	if (index == -1) {
-		ret = ocfs2_error(sb,
+		ocfs2_error(sb,
 				  "Inode %llu has an extent at cpos %u which can no longer be found\n",
 				  (unsigned long long)ino, cpos);
+		ret = -EROFS;
 		goto out;
 	}
 
@@ -3362,8 +3370,10 @@ static int ocfs2_replace_cow(struct ocfs2_cow_context *context)
 	struct ocfs2_super *osb = OCFS2_SB(inode->i_sb);
 
 	if (!ocfs2_refcount_tree(osb)) {
-		return ocfs2_error(inode->i_sb, "Inode %lu want to use refcount tree, but the feature bit is not set in the super block\n",
+		ocfs2_error(inode->i_sb,
+				   "Inode %lu want to use refcount tree, but the feature bit is not set in the super block.",
 				   inode->i_ino);
+		return -EROFS;
 	}
 
 	ocfs2_init_dealloc_ctxt(&context->dealloc);
@@ -4254,7 +4264,7 @@ static int ocfs2_reflink(struct dentry *old_dentry, struct inode *dir,
 	struct inode *inode = d_inode(old_dentry);
 	struct buffer_head *old_bh = NULL;
 	struct inode *new_orphan_inode = NULL;
-	struct ocfs2_lock_holder oh;
+	struct ocfs2_lock_holder *oh;
 
 	if (!ocfs2_refcount_tree(OCFS2_SB(inode->i_sb)))
 		return -EOPNOTSUPP;
diff --git a/fs/ocfs2/xattr.c b/fs/ocfs2/xattr.c
index 9ccd19d8f7b1..2087d93d4a0b 100644
--- a/fs/ocfs2/xattr.c
+++ b/fs/ocfs2/xattr.c
@@ -34,7 +34,7 @@
 #include "ocfs2.h"
 #include "alloc.h"
 #include "blockcheck.h"
-#include "dlmglue.h"
+#include "lockglue.h"
 #include "file.h"
 #include "symlink.h"
 #include "sysfile.h"
@@ -491,24 +491,27 @@ static int ocfs2_validate_xattr_block(struct super_block *sb,
 	 */
 
 	if (!OCFS2_IS_VALID_XATTR_BLOCK(xb)) {
-		return ocfs2_error(sb,
-				   "Extended attribute block #%llu has bad signature %.*s\n",
+		ocfs2_error(sb,
+				   "Extended attribute block #%llu has bad signature %.*s",
 				   (unsigned long long)bh->b_blocknr, 7,
 				   xb->xb_signature);
+		return -EINVAL;
 	}
 
 	if (le64_to_cpu(xb->xb_blkno) != bh->b_blocknr) {
-		return ocfs2_error(sb,
-				   "Extended attribute block #%llu has an invalid xb_blkno of %llu\n",
+		ocfs2_error(sb,
+				   "Extended attribute block #%llu has an invalid xb_blkno of %llu",
 				   (unsigned long long)bh->b_blocknr,
 				   (unsigned long long)le64_to_cpu(xb->xb_blkno));
+		return -EINVAL;
 	}
 
 	if (le32_to_cpu(xb->xb_fs_generation) != OCFS2_SB(sb)->fs_generation) {
-		return ocfs2_error(sb,
-				   "Extended attribute block #%llu has an invalid xb_fs_generation of #%u\n",
+		ocfs2_error(sb,
+				   "Extended attribute block #%llu has an invalid xb_fs_generation of #%u",
 				   (unsigned long long)bh->b_blocknr,
 				   le32_to_cpu(xb->xb_fs_generation));
+		return -EINVAL;
 	}
 
 	return 0;
@@ -722,7 +725,7 @@ static int ocfs2_xattr_extend_allocation(struct inode *inode,
 		}
 
 		prev_clusters = le32_to_cpu(vb->vb_xv->xr_clusters);
-		status = ocfs2_add_clusters_in_btree(handle,
+		status = ocfs2_add_clusters_in_btree(inode, handle,
 						     &et,
 						     &logical_start,
 						     clusters_to_add,
@@ -1032,6 +1035,13 @@ ssize_t ocfs2_listxattr(struct dentry *dentry,
 	struct ocfs2_dinode *di = NULL;
 	struct ocfs2_inode_info *oi = OCFS2_I(d_inode(dentry));
 
+	if (ocfs2_test_invalid_fs(OCFS2_SB(dentry->d_sb))) {
+		mlog(ML_ERROR,
+				"%s: OCFS2 invalid because of disk timeout.\n",
+				OCFS2_SB(dentry->d_sb)->uuid_str);
+		return -EIO;
+	}
+
 	if (!ocfs2_supports_xattr(OCFS2_SB(dentry->d_sb)))
 		return -EOPNOTSUPP;
 
@@ -1211,7 +1221,7 @@ static int ocfs2_xattr_block_get(struct inode *inode,
 	struct ocfs2_xattr_value_root *xv;
 	size_t size;
 	int ret = -ENODATA, name_offset, name_len, i;
-	int block_off;
+	int block_off = 0;
 
 	xs->bucket = ocfs2_xattr_bucket_new(inode);
 	if (!xs->bucket) {
@@ -1294,6 +1304,12 @@ int ocfs2_xattr_get_nolock(struct inode *inode,
 		.not_found = -ENODATA,
 	};
 
+	if (ocfs2_test_invalid_fs(OCFS2_SB(inode->i_sb))) {
+		mlog(ML_ERROR,
+				"%s: OCFS2 invalid because of disk timeout.\n",
+				OCFS2_SB(inode->i_sb)->uuid_str);
+		return -EIO;
+	}
 	if (!ocfs2_supports_xattr(OCFS2_SB(inode->i_sb)))
 		return -EOPNOTSUPP;
 
@@ -1325,7 +1341,7 @@ static int ocfs2_xattr_get(struct inode *inode,
 {
 	int ret, had_lock;
 	struct buffer_head *di_bh = NULL;
-	struct ocfs2_lock_holder oh;
+	struct ocfs2_lock_holder *oh;
 
 	had_lock = ocfs2_inode_lock_tracker(inode, &di_bh, 0, &oh);
 	if (had_lock < 0) {
@@ -1339,7 +1355,8 @@ static int ocfs2_xattr_get(struct inode *inode,
 
 	ocfs2_inode_unlock_tracker(inode, 0, &oh, had_lock);
 
-	brelse(di_bh);
+	if (di_bh && !had_lock)
+		brelse(di_bh);
 
 	return ret;
 }
@@ -1990,14 +2007,14 @@ static void ocfs2_xa_cleanup_value_truncate(struct ocfs2_xa_loc *loc,
 
 	if (new_clusters < orig_clusters) {
 		mlog(ML_ERROR,
-		     "Partial truncate while %s xattr %.*s.  Leaking "
+		     "ocfs2: Partial truncate while %s xattr %.*s.  Leaking "
 		     "%u clusters and removing the entry\n",
 		     what, loc->xl_entry->xe_name_len, nameval_buf,
 		     orig_clusters - new_clusters);
 		ocfs2_xa_remove_entry(loc);
 	} else if (!orig_clusters) {
 		mlog(ML_ERROR,
-		     "Unable to allocate an external value for xattr "
+		     "ocfs2: Unable to allocate an external value for xattr "
 		     "%.*s safely.  Leaking %u clusters and removing the "
 		     "entry\n",
 		     loc->xl_entry->xe_name_len, nameval_buf,
@@ -2005,7 +2022,7 @@ static void ocfs2_xa_cleanup_value_truncate(struct ocfs2_xa_loc *loc,
 		ocfs2_xa_remove_entry(loc);
 	} else if (new_clusters > orig_clusters)
 		mlog(ML_ERROR,
-		     "Unable to grow xattr %.*s safely.  %u new clusters "
+		     "ocfs2: Unable to grow xattr %.*s safely.  %u new clusters "
 		     "have been added, but the value will not be "
 		     "modified\n",
 		     loc->xl_entry->xe_name_len, nameval_buf,
@@ -2623,6 +2640,7 @@ int ocfs2_xattr_remove(struct inode *inode, struct buffer_head *di_bh)
 	spin_lock(&oi->ip_lock);
 	oi->ip_dyn_features &= ~(OCFS2_INLINE_XATTR_FL | OCFS2_HAS_XATTR_FL);
 	di->i_dyn_features = cpu_to_le16(oi->ip_dyn_features);
+	oi->ip_owner = O2NM_MAX_NODES;
 	spin_unlock(&oi->ip_lock);
 	ocfs2_update_inode_fsync_trans(handle, inode, 0);
 
@@ -2778,7 +2796,6 @@ static int ocfs2_xattr_ibody_set(struct inode *inode,
 	if (inode->i_sb->s_blocksize == OCFS2_MIN_BLOCKSIZE)
 		return -ENOSPC;
 
-	down_write(&oi->ip_alloc_sem);
 	if (!(oi->ip_dyn_features & OCFS2_INLINE_XATTR_FL)) {
 		ret = ocfs2_xattr_ibody_init(inode, xs->inode_bh, ctxt);
 		if (ret) {
@@ -2799,8 +2816,6 @@ static int ocfs2_xattr_ibody_set(struct inode *inode,
 	xs->here = loc.xl_entry;
 
 out:
-	up_write(&oi->ip_alloc_sem);
-
 	return ret;
 }
 
@@ -3507,7 +3522,9 @@ int ocfs2_xattr_set_handle(handle_t *handle,
 			goto cleanup;
 	}
 
+	down_write(&OCFS2_I(inode)->ip_alloc_sem);
 	ret = __ocfs2_xattr_set_handle(inode, di, &xi, &xis, &xbs, &ctxt);
+	up_write(&OCFS2_I(inode)->ip_alloc_sem);
 
 cleanup:
 	up_write(&OCFS2_I(inode)->ip_xattr_sem);
@@ -3533,12 +3550,13 @@ int ocfs2_xattr_set(struct inode *inode,
 {
 	struct buffer_head *di_bh = NULL;
 	struct ocfs2_dinode *di;
+	struct ocfs2_inode_info *oi = OCFS2_I(inode);
 	int ret, credits, had_lock, ref_meta = 0, ref_credits = 0;
 	struct ocfs2_super *osb = OCFS2_SB(inode->i_sb);
 	struct inode *tl_inode = osb->osb_tl_inode;
 	struct ocfs2_xattr_set_ctxt ctxt = { NULL, NULL, NULL, };
 	struct ocfs2_refcount_tree *ref_tree = NULL;
-	struct ocfs2_lock_holder oh;
+	struct ocfs2_lock_holder *oh;
 
 	struct ocfs2_xattr_info xi = {
 		.xi_name_index = name_index,
@@ -3578,7 +3596,7 @@ int ocfs2_xattr_set(struct inode *inode,
 	xis.inode_bh = xbs.inode_bh = di_bh;
 	di = (struct ocfs2_dinode *)di_bh->b_data;
 
-	down_write(&OCFS2_I(inode)->ip_xattr_sem);
+	down_write(&oi->ip_xattr_sem);
 	/*
 	 * Scan inode and external block to find the same name
 	 * extended attribute and collect search information.
@@ -3629,9 +3647,11 @@ int ocfs2_xattr_set(struct inode *inode,
 	}
 	inode_unlock(tl_inode);
 
+	down_write(&oi->ip_alloc_sem);
 	ret = ocfs2_init_xattr_set_ctxt(inode, di, &xi, &xis,
 					&xbs, &ctxt, ref_meta, &credits);
 	if (ret) {
+		up_write(&oi->ip_alloc_sem);
 		mlog_errno(ret);
 		goto cleanup;
 	}
@@ -3658,11 +3678,12 @@ int ocfs2_xattr_set(struct inode *inode,
 	if (ocfs2_dealloc_has_cluster(&ctxt.dealloc))
 		ocfs2_schedule_truncate_log_flush(osb, 1);
 	ocfs2_run_deallocs(osb, &ctxt.dealloc);
+	up_write(&oi->ip_alloc_sem);
 
 cleanup:
 	if (ref_tree)
 		ocfs2_unlock_refcount_tree(osb, ref_tree, 1);
-	up_write(&OCFS2_I(inode)->ip_xattr_sem);
+	up_write(&oi->ip_xattr_sem);
 	if (!value && !ret) {
 		ret = ocfs2_try_remove_refcount_tree(inode, di_bh);
 		if (ret)
@@ -3707,10 +3728,11 @@ static int ocfs2_xattr_get_rec(struct inode *inode,
 		el = &eb->h_list;
 
 		if (el->l_tree_depth) {
-			ret = ocfs2_error(inode->i_sb,
+			ocfs2_error(inode->i_sb,
 					  "Inode %lu has non zero tree depth in xattr tree block %llu\n",
 					  inode->i_ino,
 					  (unsigned long long)eb_bh->b_blocknr);
+			ret = -EROFS;
 			goto out;
 		}
 	}
@@ -3725,10 +3747,12 @@ static int ocfs2_xattr_get_rec(struct inode *inode,
 	}
 
 	if (!e_blkno) {
-		ret = ocfs2_error(inode->i_sb, "Inode %lu has bad extent record (%u, %u, 0) in xattr\n",
-				  inode->i_ino,
-				  le32_to_cpu(rec->e_cpos),
-				  ocfs2_rec_clusters(el, rec));
+		ocfs2_error(inode->i_sb,
+					  "Inode %lu has bad extent record (%u, %u, 0) in xattr",
+					  inode->i_ino,
+					  le32_to_cpu(rec->e_cpos),
+					  ocfs2_rec_clusters(el, rec));
+		ret = -EROFS;
 		goto out;
 	}
 
@@ -4277,7 +4301,6 @@ static int ocfs2_xattr_create_index_block(struct inode *inode,
 	u32 bit_off, len;
 	u64 blkno;
 	handle_t *handle = ctxt->handle;
-	struct ocfs2_inode_info *oi = OCFS2_I(inode);
 	struct buffer_head *xb_bh = xs->xattr_bh;
 	struct ocfs2_xattr_block *xb =
 			(struct ocfs2_xattr_block *)xb_bh->b_data;
@@ -4290,13 +4313,6 @@ static int ocfs2_xattr_create_index_block(struct inode *inode,
 	BUG_ON(xb_flags & OCFS2_XATTR_INDEXED);
 	BUG_ON(!xs->bucket);
 
-	/*
-	 * XXX:
-	 * We can use this lock for now, and maybe move to a dedicated mutex
-	 * if performance becomes a problem later.
-	 */
-	down_write(&oi->ip_alloc_sem);
-
 	ret = ocfs2_journal_access_xb(handle, INODE_CACHE(inode), xb_bh,
 				      OCFS2_JOURNAL_ACCESS_WRITE);
 	if (ret) {
@@ -4358,8 +4374,6 @@ static int ocfs2_xattr_create_index_block(struct inode *inode,
 	ocfs2_journal_dirty(handle, xb_bh);
 
 out:
-	up_write(&oi->ip_alloc_sem);
-
 	return ret;
 }
 
@@ -5527,8 +5541,9 @@ static int ocfs2_check_xattr_bucket_collision(struct inode *inode,
 
 	if (xh->xh_entries[le16_to_cpu(xh->xh_count) - 1].xe_name_hash ==
 	    xh->xh_entries[0].xe_name_hash) {
-		mlog(ML_ERROR, "Too much hash collision in xattr bucket %llu, "
-		     "hash = %u\n",
+		mlog(ML_ERROR,
+		     "%s: Too much hash collision in xattr bucket %llu, hash = %u\n",
+		     OCFS2_SB(inode->i_sb)->uuid_str,
 		     (unsigned long long)bucket_blkno(bucket),
 		     le32_to_cpu(xh->xh_entries[0].xe_name_hash));
 		return -ENOSPC;
@@ -7240,6 +7255,7 @@ int ocfs2_init_security_and_acl(struct inode *dir,
 /*
  * 'security' attributes support
  */
+
 static int ocfs2_xattr_security_get(const struct xattr_handler *handler,
 				    struct dentry *unused, struct inode *inode,
 				    const char *name, void *buffer, size_t size)
@@ -7253,6 +7269,7 @@ static int ocfs2_xattr_security_set(const struct xattr_handler *handler,
 				    const char *name, const void *value,
 				    size_t size, int flags)
 {
+
 	return ocfs2_xattr_set(inode, OCFS2_XATTR_INDEX_SECURITY,
 			       name, value, size, flags);
 }
@@ -7325,6 +7342,7 @@ static int ocfs2_xattr_trusted_set(const struct xattr_handler *handler,
 				   const char *name, const void *value,
 				   size_t size, int flags)
 {
+
 	return ocfs2_xattr_set(inode, OCFS2_XATTR_INDEX_TRUSTED,
 			       name, value, size, flags);
 }
@@ -7346,22 +7364,147 @@ static int ocfs2_xattr_user_get(const struct xattr_handler *handler,
 
 	if (osb->s_mount_opt & OCFS2_MOUNT_NOUSERXATTR)
 		return -EOPNOTSUPP;
+
+	if (ocfs2_test_invalid_fs(osb)) {
+		mlog(ML_ERROR,
+				"%s: OCFS2 invalid because of disk timeout.\n",
+				osb->uuid_str);
+		return -EIO;
+	}
+
 	return ocfs2_xattr_get(inode, OCFS2_XATTR_INDEX_USER, name,
 			       buffer, size);
 }
 
+static int ocfs2_xattr_compatible(struct ocfs2_super *osb,
+		struct inode *inode, const char *xattr_name)
+{
+	int ret = 0;
+	char kvalue[XATTR_OWNER_VALUE_LEN + 1] = {0};
+	struct ocfs2_inode_info *oi = OCFS2_I(inode);
+
+	if (XATTR_NAME_IS_OWNER(xattr_name)) {
+		ret = ocfs2_inode_get_share_vol(inode);
+		if (ret < 0)
+			mlog_errno(ret);
+		else {
+			spin_lock(&oi->ip_lock);
+			ret = (oi->ip_shared_vol == 1) ? 0 : 1;
+			spin_unlock(&oi->ip_lock);
+		}
+	} else {
+		ret = ocfs2_xattr_get(inode, OCFS2_XATTR_INDEX_USER,
+				XATTR_OWNER_NAME, kvalue, sizeof(kvalue));
+		if (ret < 0 && ret != -ENODATA && ret != -EOPNOTSUPP)
+			mlog_errno(ret);
+		else
+			ret = (ret == -ENODATA || ret == -EOPNOTSUPP) ? 1 : 0;
+	}
+
+	return ret;
+}
+
+static int ocfs2_xattr_sharedvol_and_owner_set(struct dentry *dentry, const char *name,
+		const void *value, size_t size, int flags)
+{
+	int ret = 0;
+	struct ocfs2_super *osb = OCFS2_SB(dentry->d_sb);
+	struct inode *inode = d_inode(dentry);
+	struct ocfs2_inode_info *oi = OCFS2_I(inode);
+	char kvalue[XATTR_OWNER_VALUE_LEN + 1] = {0};
+
+	if (!S_ISREG(inode->i_mode)) {
+		mlog(ML_ERROR,
+				"%s: not support setting %s for non-regular file %pd\n",
+				osb->uuid_str, name, dentry);
+		return -EINVAL;
+	}
+
+	ret = ocfs2_xattr_compatible(osb, inode, name);
+	if (ret < 0) {
+		mlog_errno(ret);
+		return ret;
+	} else if (ret == 0) {
+		mlog(ML_ERROR,
+				"%s: the %s attribute has been set, can not set the %s attribute\n",
+				osb->uuid_str,
+				XATTR_NAME_IS_OWNER(name) ?
+				XATTR_SHARED_VOL_NAME : XATTR_OWNER_NAME,
+				name);
+		ret = -EINVAL;
+		return ret;
+	}
+
+	if (flags & XATTR_REPLACE) {
+		mlog(ML_NOTICE, "%s: clear %s node %u for %pd\n",
+				osb->uuid_str, name, osb->node_num,
+				dentry);
+		ret = ocfs2_xattr_set(inode, OCFS2_XATTR_INDEX_USER,
+				name, value, size, flags);
+		if (!ret) {
+			spin_lock(&oi->ip_lock);
+			if (XATTR_NAME_IS_OWNER(name))
+				oi->ip_owner = O2NM_MAX_NODES;
+			else
+				oi->ip_shared_vol = 0;
+			spin_unlock(&oi->ip_lock);
+		} else
+			mlog_errno(ret);
+	} else {
+		mlog(ML_NOTICE, "%s: set %s node %u for %pd\n",
+				osb->uuid_str, name, osb->node_num,
+				dentry);
+		if (XATTR_NAME_IS_OWNER(name)) {
+			snprintf(kvalue, XATTR_OWNER_VALUE_LEN + 1, "%u", osb->node_num);
+			ret = ocfs2_xattr_set(inode, OCFS2_XATTR_INDEX_USER,
+					name, kvalue, strlen(kvalue), flags);
+		} else
+			ret = ocfs2_xattr_set(inode, OCFS2_XATTR_INDEX_USER,
+					name, value, size, flags);
+
+		if (!ret) {
+			spin_lock(&oi->ip_lock);
+			if (XATTR_NAME_IS_OWNER(name))
+				oi->ip_owner = osb->node_num;
+			else
+				oi->ip_shared_vol = 1;
+			spin_unlock(&oi->ip_lock);
+		} else
+			mlog_errno(ret);
+	}
+
+	return ret;
+}
+
 static int ocfs2_xattr_user_set(const struct xattr_handler *handler,
 				struct dentry *unused, struct inode *inode,
 				const char *name, const void *value,
 				size_t size, int flags)
 {
+	int ret;
 	struct ocfs2_super *osb = OCFS2_SB(inode->i_sb);
 
 	if (osb->s_mount_opt & OCFS2_MOUNT_NOUSERXATTR)
 		return -EOPNOTSUPP;
 
-	return ocfs2_xattr_set(inode, OCFS2_XATTR_INDEX_USER,
-			       name, value, size, flags);
+	if (flags & XATTR_REPLACE)
+		mlog(ML_NOTICE, "%s: clear xattr from %pd, name %s\n",
+			osb->uuid_str, unused, name);
+	else
+		mlog(ML_NOTICE, "%s: set xattr to %pd, name %s, value %.*s\n",
+			osb->uuid_str, unused, name,
+			(int)size, (char *)value);
+
+	if (XATTR_NAME_IS_OWNER(name) ||
+			XATTR_NAME_IS_SHARED_VOL(name)) {
+		ret = ocfs2_xattr_sharedvol_and_owner_set(unused, name,
+				value, size, flags);
+		if (ret < 0)
+			mlog_errno(ret);
+		return ret;
+	} else
+		return ocfs2_xattr_set(inode, OCFS2_XATTR_INDEX_USER,
+				name, value, size, flags);
 }
 
 const struct xattr_handler ocfs2_xattr_user_handler = {
@@ -7369,3 +7512,55 @@ const struct xattr_handler ocfs2_xattr_user_handler = {
 	.get	= ocfs2_xattr_user_get,
 	.set	= ocfs2_xattr_user_set,
 };
+
+int ocfs2_getxattr_owner(struct inode *inode,
+		struct buffer_head *di_bh,
+		void *buffer,
+		size_t buffer_size)
+{
+	int ret;
+	struct ocfs2_super *osb = OCFS2_SB(inode->i_sb);
+
+	if (osb->s_mount_opt & OCFS2_MOUNT_NOUSERXATTR)
+		return -EOPNOTSUPP;
+
+	down_read(&OCFS2_I(inode)->ip_xattr_sem);
+	ret = ocfs2_xattr_get_nolock(inode, di_bh, OCFS2_XATTR_INDEX_USER,
+				     XATTR_OWNER_NAME, buffer, buffer_size);
+	up_read(&OCFS2_I(inode)->ip_xattr_sem);
+	if (ret < 0 && ret != -ENODATA && ret != -EOPNOTSUPP)
+		mlog_errno(ret);
+
+	return ret;
+}
+
+int ocfs2_inode_get_share_vol(struct inode *inode)
+{
+	int ret;
+	struct ocfs2_inode_info *oi = OCFS2_I(inode);
+	struct ocfs2_super *osb = OCFS2_SB(inode->i_sb);
+	char buf[XATTR_SHARED_VOL_VALUE_LEN + 1] = {0};
+
+	if (osb->s_mount_opt & OCFS2_MOUNT_NOUSERXATTR)
+		return 0;
+
+	ret = ocfs2_xattr_get(inode, OCFS2_XATTR_INDEX_USER,
+			XATTR_SHARED_VOL_NAME, buf, sizeof(buf));
+	if (ret < 0) {
+		if (ret == -ENODATA || ret == -EOPNOTSUPP) {
+			spin_lock(&oi->ip_lock);
+			oi->ip_shared_vol = 0;
+			spin_unlock(&oi->ip_lock);
+			ret = 0;
+		} else {
+			mlog_errno(ret);
+		}
+		return ret;
+	}
+
+	spin_lock(&oi->ip_lock);
+	oi->ip_shared_vol = XATTR_SHARED_VOL_VALUE_IS_YES(buf) ? 1 : 0;
+	spin_unlock(&oi->ip_lock);
+
+	return 0;
+}
diff --git a/fs/ocfs2/xattr.h b/fs/ocfs2/xattr.h
index 9c80382da1f5..e5693b3e3627 100644
--- a/fs/ocfs2/xattr.h
+++ b/fs/ocfs2/xattr.h
@@ -87,4 +87,22 @@ int ocfs2_reflink_xattrs(struct inode *old_inode,
 int ocfs2_init_security_and_acl(struct inode *dir,
 				struct inode *inode,
 				const struct qstr *qstr);
+
+#define XATTR_SHARED_VOL_NAME "sharedvol"
+#define XATTR_SHARED_VOL_VALUE "yes"
+#define XATTR_SHARED_VOL_VALUE_LEN 3
+#define XATTR_SHARED_VOL_VALUE_IS_YES(value) (!strncmp(value, XATTR_SHARED_VOL_VALUE, \
+		XATTR_SHARED_VOL_VALUE_LEN + 1))
+#define XATTR_NAME_IS_SHARED_VOL(name) (!strncmp(name, XATTR_SHARED_VOL_NAME, \
+		strlen(XATTR_SHARED_VOL_NAME) + 1))
+
+#define XATTR_OWNER_NAME "owner"
+#define XATTR_OWNER_VALUE_LEN 4
+#define XATTR_NAME_IS_OWNER(name) (!strncmp(name, XATTR_OWNER_NAME, strlen(XATTR_OWNER_NAME) + 1))
+
+int ocfs2_getxattr_owner(struct inode *inode,
+		struct buffer_head *di_bh,
+		void *buffer,
+		size_t buffer_size);
+int ocfs2_inode_get_share_vol(struct inode *inode);
 #endif /* OCFS2_XATTR_H */
-- 
2.17.1

